
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.1.6">
    
    
      
        <title>Bonus code - Enrichment analysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.875de78c.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.f1a3b89f.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    <script>function __prefix(e){return new URL("..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#bonus-code" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Enrichment analysis" class="md-header__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Enrichment analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Bonus code
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
      <div class="md-header__source">
        
<a href="https://github.com/sib-swiss/enrichment-analysis-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Enrichment analysis" class="md-nav__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Enrichment analysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sib-swiss/enrichment-analysis-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../materials/" class="md-nav__link">
        Materials
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../exercises/" class="md-nav__link">
        Exercises
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Bonus code
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Bonus code
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bonus-code" class="md-nav__link">
    Bonus code
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-label-conversions" class="md-nav__link">
    Gene label conversions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#msigdbr-package" class="md-nav__link">
    msigdbr package
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-a-heatmap-of-p-values" class="md-nav__link">
    Code for a heatmap of p-values
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../links/" class="md-nav__link">
        Useful links
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bonus-code" class="md-nav__link">
    Bonus code
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-label-conversions" class="md-nav__link">
    Gene label conversions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#msigdbr-package" class="md-nav__link">
    msigdbr package
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-a-heatmap-of-p-values" class="md-nav__link">
    Code for a heatmap of p-values
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/sib-swiss/enrichment-analysis-training/edit/master/docs/bonus_code.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Bonus code</h1>
                
                <h2 id="bonus-code"><strong>Bonus code</strong> <img alt="ðŸ¥‚" class="twemoji" src="https://twemoji.maxcdn.com/v/latest/svg/1f942.svg" title=":champagne_glass:" /></h2>
<p>The following code was added thanks to questions from course participants of past sessions. They might be useful for you too.</p>
<h2 id="gene-label-conversions">Gene label conversions</h2>
<p>It is often useful to convert different types of gene labels even further. Here is an example for gene label conversion 
and gene information extraction using <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>. A lot of information for each gene can be obtained, such as chromosome location,
description, biotype, or the symbols of mouse homologs of human genes, etc.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># install and load the package:</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;biomaRt&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">biomaRt</span><span class="p">)</span>
<span class="c1"># list the available options</span>
<span class="nf">listEnsembl</span><span class="p">()</span>

<span class="n">ensembl</span> <span class="o">&lt;-</span> <span class="nf">useEnsembl</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s">&quot;genes&quot;</span><span class="p">)</span>

<span class="c1"># List the available species and reference genomes:</span>
<span class="n">datasets</span> <span class="o">&lt;-</span> <span class="nf">listDatasets</span><span class="p">(</span><span class="n">ensembl</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">datasets</span><span class="p">)</span>
<span class="c1"># search for human dataset and genome version:</span>
<span class="n">datasets</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;sapiens&quot;</span><span class="p">,</span> <span class="n">datasets</span><span class="o">$</span><span class="n">dataset</span><span class="p">),]</span>
<span class="c1"># 80 hsapiens_gene_ensembl Human genes (GRCh38.p13) GRCh38.p13</span>

<span class="c1"># To obtain information on human genes, first the data specific to human has to be accessed from the online Ensembl repository</span>
<span class="n">ensembl</span> <span class="o">&lt;-</span> <span class="nf">useEnsembl</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s">&quot;genes&quot;</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>

<span class="c1"># you can extract gene biotype, description, wiki gene description,</span>
<span class="c1"># chromosome location of genes, even strand information etc</span>
<span class="c1"># using the listAttributes function allows you to view the type of information per gene you can extract.</span>
<span class="n">attributes</span> <span class="o">&lt;-</span> <span class="nf">listAttributes</span><span class="p">(</span><span class="n">ensembl</span><span class="p">)</span>
<span class="n">attributes</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">15</span><span class="p">,]</span>

<span class="c1"># Now we extract symbol (external_gene_name), description, gene_biotype (eg whether it is a protein coding or other type of gene):</span>
<span class="n">biomart_gene_info</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span> <span class="s">&quot;external_gene_name&quot;</span><span class="p">,</span> <span class="s">&quot;description&quot;</span><span class="p">,</span> <span class="s">&quot;gene_biotype&quot;</span><span class="p">,</span> <span class="s">&quot;wikigene_description&quot;</span><span class="p">),</span>
                           <span class="n">filter</span><span class="o">=</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span>
                           <span class="n">values</span><span class="o">=</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">ensembl_gene_id</span><span class="p">,</span>
                           <span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
<span class="c1"># check the structure of the resulting data frame:</span>
<span class="nf">head</span><span class="p">(</span><span class="n">biomart_gene_info</span><span class="p">)</span>

<span class="c1"># search for the oncogene key word in the description:</span>
<span class="n">biomart_gene_info</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;oncogene&quot;</span><span class="p">,</span> <span class="n">biomart_gene_info</span><span class="o">$</span><span class="n">description</span><span class="p">),]</span>
<span class="n">biomart_gene_info</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;tumor protein&quot;</span><span class="p">,</span> <span class="n">biomart_gene_info</span><span class="o">$</span><span class="n">description</span><span class="p">),]</span>

<span class="c1"># search for genes that have TP53 in their gene symbol:</span>
<span class="n">biomart_gene_info</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;TP53&quot;</span><span class="p">,</span> <span class="n">biomart_gene_info</span><span class="o">$</span><span class="n">external_gene_name</span><span class="p">),]</span>
</code></pre></div>
<p>For conversion of human gene symbols to mouse homologs for example (or vice versa if you provide mouse genes as the &ldquo;values&rdquo; arguments
                                                                    and &ldquo;hsapiens_homolog_associated_gene_name&rdquo; in the list of the attributes argument), you can also use biomaRt.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Convert human genes to mouse homologs</span>
<span class="n">ensembl_human_to_mouse</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s">&quot;external_gene_name&quot;</span><span class="p">,</span> <span class="s">&quot;mmusculus_homolog_associated_gene_name&quot;</span><span class="p">),</span>
                                <span class="n">filter</span><span class="o">=</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span>
                                <span class="n">values</span><span class="o">=</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">ensembl_gene_id</span><span class="p">,</span>
                                <span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">ensembl_human_to_mouse</span><span class="p">)</span>
<span class="c1"># ensembl_gene_id external_gene_name mmusculus_homolog_associated_gene_name</span>
<span class="c1"># 1 ENSG00000000003             TSPAN6                                 Tspan6</span>
<span class="c1"># 2 ENSG00000000419               DPM1                                   Dpm1</span>
<span class="c1"># 3 ENSG00000000419               DPM1                                Gm20716</span>
<span class="c1"># 4 ENSG00000000457              SCYL3                                  Scyl3</span>
<span class="c1"># 5 ENSG00000000460           C1orf112                               BC055324</span>
<span class="c1"># 6 ENSG00000000938                FGR                                    Fgr</span>
</code></pre></div>
<h2 id="msigdbr-package">msigdbr package</h2>
<p>The <a href="https://cran.r-project.org/web/packages/msigdbr/index.html">msigdbr</a> package hosted on CRAN allows to access gene set collections hosted on MSigDB directly within R. Check out its vignette to view how to download collections for other species such as mouse, etc. The function <code>msigdbr_species()</code> allows you to list available species. </p>
<p>For example, to download the Hallmark collection with human gene symbols within R:
<div class="highlight"><pre><span></span><code><span class="n">gmt</span> <span class="o">&lt;-</span> <span class="n">msigdbr</span><span class="o">::</span><span class="nf">msigdbr</span><span class="p">(</span><span class="n">species</span> <span class="o">=</span> <span class="s">&quot;human&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="s">&quot;H&quot;</span><span class="p">)</span>

<span class="c1"># Create the 2 column-format (TERM2GENE argument) required by clusterProfiler:</span>
<span class="n">h_gmt</span> <span class="o">&lt;-</span> <span class="n">gmt</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs_name&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">)]</span>
</code></pre></div></p>
<h2 id="code-for-a-heatmap-of-p-values">Code for a heatmap of p-values</h2>
<p>For heatmaps, ggplot2 can also be used. Here is an example for a heatmap of the p-values of 6 different gene sets (gs), and the
p-value for each gene set was calculated in two comparisons, so we compared the enrichment in genes differentially expressed
between NK cells and Th cells, and between NK cells and CD8 T cells (note that this is a dummy example just to show you the 
                                                                     example of the code, it is not based on real RNA seq data).
Before using ggplot2, you need to create a dataframe that contains a column with the gene set identities, a column with the 
name of the cell type comparisons, and a column with the p-value of each gene set in each comparison. So a specific format is 
required for ggplot2.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Create a data frame that contains the p-value for every gene set for every </span>
<span class="c1"># cell type comparison. You need to include the values also for the non-significant</span>
<span class="c1"># gene sets. If you use function of the clusterProfiler package, you will usually</span>
<span class="c1"># obtain results in the @result slot only for significant gene sets. But if you need</span>
<span class="c1"># p-values also for the non-significant gene sets, you can change the argument</span>
<span class="c1"># pvalueCutoff = 1 in the gseGO function (and other functions of the clusterProfiler</span>
<span class="c1"># package)</span>

<span class="c1"># For the example, create a dummy data frame with the list of gene sets (gs), the</span>
<span class="c1"># 2 cell type comparisons, and the p-value for each gene set in each comparison:</span>
<span class="n">ora_to_plot</span><span class="o">&lt;-</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">cbind</span><span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs1&quot;</span><span class="p">,</span> <span class="s">&quot;gs2&quot;</span><span class="p">,</span><span class="s">&quot;gs3&quot;</span><span class="p">,</span><span class="s">&quot;gs4&quot;</span><span class="p">,</span> <span class="s">&quot;gs5&quot;</span><span class="p">,</span> <span class="s">&quot;gs6&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;gs1&quot;</span><span class="p">,</span> <span class="s">&quot;gs2&quot;</span><span class="p">,</span><span class="s">&quot;gs3&quot;</span><span class="p">,</span><span class="s">&quot;gs4&quot;</span><span class="p">,</span> <span class="s">&quot;gs5&quot;</span><span class="p">,</span> <span class="s">&quot;gs6&quot;</span><span class="p">),</span>
                                 <span class="n">comparison</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;NK_vs_Th&quot;</span><span class="p">,</span><span class="m">6</span><span class="p">),</span> <span class="nf">rep</span><span class="p">(</span><span class="s">&quot;NK_vs_CD8&quot;</span><span class="p">,</span><span class="m">6</span><span class="p">)),</span>
                                 <span class="n">p_val</span><span class="o">=</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0.8</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span> <span class="m">0.054</span><span class="p">,</span> <span class="m">0.00001</span><span class="p">,</span> <span class="m">0.0003</span><span class="p">,</span>
                                                    <span class="m">0.0002</span><span class="p">,</span> <span class="m">0.004</span><span class="p">,</span> <span class="m">0.001</span><span class="p">,</span><span class="m">0.01</span><span class="p">,</span> <span class="m">0.85</span><span class="p">,</span><span class="m">0.9</span><span class="p">))))</span>

<span class="c1"># transform the p-val to -log10(p-val):</span>
<span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="o">&lt;--</span><span class="nf">log10</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="o">$</span><span class="n">p_val</span><span class="p">))</span>
<span class="c1"># set the none-significant p-values to 0 so that they appear grey in the heatmap:</span>
<span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="o">&lt;-</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="o">&lt;</span><span class="p">(</span><span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="m">0.05</span><span class="p">)),</span> <span class="m">0</span><span class="p">,</span> <span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="p">)</span>
<span class="nf">range</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="p">)</span> <span class="c1"># [1]  0 5</span>

<span class="nf">head</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="m">12</span><span class="p">)</span>
<span class="c1">#     ID comparison p_val log10_p_val</span>
<span class="c1"># 1  gs1   NK_vs_Th   0.8    0.000000</span>
<span class="c1"># 2  gs2   NK_vs_Th   0.9    0.000000</span>
<span class="c1"># 3  gs3   NK_vs_Th   0.6    0.000000</span>
<span class="c1"># 4  gs4   NK_vs_Th 0.054    0.000000</span>
<span class="c1"># 5  gs5   NK_vs_Th 1e-05    5.000000</span>
<span class="c1"># 6  gs6   NK_vs_Th 3e-04    3.522879</span>
<span class="c1"># 7  gs1  NK_vs_CD8 2e-04    3.698970</span>
<span class="c1"># 8  gs2  NK_vs_CD8 0.004    2.397940</span>
<span class="c1"># 9  gs3  NK_vs_CD8 0.001    3.000000</span>
<span class="c1"># 10 gs4  NK_vs_CD8  0.01    2.000000</span>
<span class="c1"># 11 gs5  NK_vs_CD8  0.85    0.000000</span>
<span class="c1"># 12 gs6  NK_vs_CD8   0.9    0.000000</span>

<span class="c1"># create a palette of colors based on the Plasma palette (grDevices package)</span>
<span class="nf">plot</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="nf">hcl.colors</span><span class="p">(</span><span class="m">20</span><span class="p">,</span><span class="s">&quot;Plasma&quot;</span><span class="p">),</span> <span class="n">pch</span><span class="o">=</span><span class="m">15</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">3</span><span class="p">)</span>
<span class="nf">hcl.colors</span><span class="p">(</span><span class="m">9</span><span class="p">,</span><span class="s">&quot;Plasma&quot;</span><span class="p">)</span><span class="c1"># &quot;RdYlGn&quot;)</span>
<span class="n">breaks</span><span class="o">&lt;-</span><span class="nf">seq</span><span class="p">(</span><span class="n">from</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">to</span><span class="o">=</span><span class="m">5</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="m">0.5</span> <span class="p">)</span>

<span class="n">color</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;grey78&quot;</span><span class="p">,</span> <span class="nf">rev</span><span class="p">(</span><span class="nf">hcl.colors</span><span class="p">(</span><span class="m">9</span><span class="p">,</span><span class="s">&quot;Plasma&quot;</span><span class="p">)))</span>

<span class="c1"># create plot and export as png:</span>
<span class="n">p</span><span class="o">&lt;-</span><span class="nf">ggplot</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">comparison</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span> <span class="n">log10_p_val</span><span class="p">))</span> <span class="o">+</span> 
  <span class="nf">geom_tile</span><span class="p">(</span><span class="n">color</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_fill_gradientn</span><span class="p">(</span><span class="n">breaks</span><span class="o">=</span> <span class="n">breaks</span><span class="p">,</span> 
                       <span class="n">colors</span> <span class="o">=</span> <span class="n">color</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme_bw</span><span class="p">()</span> 
<span class="nf">ggsave</span><span class="p">(</span><span class="n">plot</span> <span class="o">=</span> <span class="n">p</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s">&quot;heatmap_p_value_ORA.png&quot;</span><span class="p">,</span> 
       <span class="n">device</span><span class="o">=</span><span class="s">&quot;png&quot;</span><span class="p">,</span>
       <span class="n">width</span> <span class="o">=</span> <span class="m">4</span><span class="p">,</span><span class="n">height</span> <span class="o">=</span> <span class="m">6</span><span class="p">)</span>
</code></pre></div>
<p>You will obtain the following heatmap:</p>
<figure>
  <img src="../assets/images/heatmap_p_value_ORA.png" width="300"/>
  </figure>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        <a href="../exercises/" class="md-footer__link md-footer__link--prev" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Exercises
            </div>
          </div>
        </a>
      
      
        <a href="../links/" class="md-footer__link md-footer__link--next" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Useful links
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../assets/javascripts/workers/search.d351de03.min.js", "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.34eae1b6.min.js"></script>
      
    
  </body>
</html>