
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.1.6">
    
    
      
        <title>Bonus code - Enrichment analysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.875de78c.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.f1a3b89f.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    <script>function __prefix(e){return new URL("..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#bonus-code" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Enrichment analysis" class="md-header__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Enrichment analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Bonus code
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
      <div class="md-header__source">
        
<a href="https://github.com/sib-swiss/enrichment-analysis-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Enrichment analysis" class="md-nav__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Enrichment analysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sib-swiss/enrichment-analysis-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../materials/" class="md-nav__link">
        Materials
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../exercises/" class="md-nav__link">
        Exercises
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Bonus code
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Bonus code
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bonus-code" class="md-nav__link">
    Bonus code
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-label-conversions" class="md-nav__link">
    Gene label conversions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#msigdbr-package" class="md-nav__link">
    msigdbr package
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-barplots-with-ggplot2" class="md-nav__link">
    Code for barplots with ggplot2
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-a-heatmap-of-p-values-with-ggplot2" class="md-nav__link">
    Code for a heatmap of p-values with ggplot2
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-a-lollipop-plot-with-ggplot2" class="md-nav__link">
    Code for a lollipop plot with ggplot2
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#small-information-on-a-typical-rnaseq-workflow" class="md-nav__link">
    Small information on a typical RNAseq workflow
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-for-de-analysis-with-deseq2" class="md-nav__link">
    Example for DE analysis with DESeq2
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../links/" class="md-nav__link">
        Useful links
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bonus-code" class="md-nav__link">
    Bonus code
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-label-conversions" class="md-nav__link">
    Gene label conversions
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#msigdbr-package" class="md-nav__link">
    msigdbr package
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-barplots-with-ggplot2" class="md-nav__link">
    Code for barplots with ggplot2
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-a-heatmap-of-p-values-with-ggplot2" class="md-nav__link">
    Code for a heatmap of p-values with ggplot2
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#code-for-a-lollipop-plot-with-ggplot2" class="md-nav__link">
    Code for a lollipop plot with ggplot2
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#small-information-on-a-typical-rnaseq-workflow" class="md-nav__link">
    Small information on a typical RNAseq workflow
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-for-de-analysis-with-deseq2" class="md-nav__link">
    Example for DE analysis with DESeq2
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/sib-swiss/enrichment-analysis-training/edit/master/docs/bonus_code.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Bonus code</h1>
                
                <h2 id="bonus-code"><strong>Bonus code</strong></h2>
<p>The following code was added thanks to questions from course participants of past sessions. They might be useful for you too.</p>
<h2 id="gene-label-conversions">Gene label conversions</h2>
<p>It is often useful to convert different types of gene labels even further. Here is an example for gene label conversion 
and gene information extraction using <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>. A lot of information for each gene can be obtained, such as chromosome location,
description, biotype, or the symbols of mouse homologs of human genes, etc.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># install and load the package:</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;biomaRt&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">biomaRt</span><span class="p">)</span>
<span class="c1"># list the available options</span>
<span class="nf">listEnsembl</span><span class="p">()</span>

<span class="n">ensembl</span> <span class="o">&lt;-</span> <span class="nf">useEnsembl</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s">&quot;genes&quot;</span><span class="p">)</span>

<span class="c1"># List the available species and reference genomes:</span>
<span class="n">datasets</span> <span class="o">&lt;-</span> <span class="nf">listDatasets</span><span class="p">(</span><span class="n">ensembl</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">datasets</span><span class="p">)</span>
<span class="c1"># search for human dataset and genome version:</span>
<span class="n">datasets</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;sapiens&quot;</span><span class="p">,</span> <span class="n">datasets</span><span class="o">$</span><span class="n">dataset</span><span class="p">),]</span>
<span class="c1"># 80 hsapiens_gene_ensembl Human genes (GRCh38.p13) GRCh38.p13</span>

<span class="c1"># To obtain information on human genes, first the data specific to human has to be accessed from the online Ensembl repository</span>
<span class="n">ensembl</span> <span class="o">&lt;-</span> <span class="nf">useEnsembl</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s">&quot;genes&quot;</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>

<span class="c1"># you can extract gene biotype, description, wiki gene description,</span>
<span class="c1"># chromosome location of genes, even strand information etc</span>
<span class="c1"># using the listAttributes function allows you to view the type of information per gene you can extract.</span>
<span class="n">attributes</span> <span class="o">&lt;-</span> <span class="nf">listAttributes</span><span class="p">(</span><span class="n">ensembl</span><span class="p">)</span>
<span class="n">attributes</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">15</span><span class="p">,]</span>

<span class="c1"># Now we extract symbol (external_gene_name), description, gene_biotype (eg whether it is a protein coding or other type of gene):</span>
<span class="n">biomart_gene_info</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span> <span class="s">&quot;external_gene_name&quot;</span><span class="p">,</span> <span class="s">&quot;description&quot;</span><span class="p">,</span> <span class="s">&quot;gene_biotype&quot;</span><span class="p">,</span> <span class="s">&quot;wikigene_description&quot;</span><span class="p">),</span>
                           <span class="n">filter</span><span class="o">=</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span>
                           <span class="n">values</span><span class="o">=</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">ensembl_gene_id</span><span class="p">,</span>
                           <span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
<span class="c1"># check the structure of the resulting data frame:</span>
<span class="nf">head</span><span class="p">(</span><span class="n">biomart_gene_info</span><span class="p">)</span>

<span class="c1"># search for the oncogene key word in the description:</span>
<span class="n">biomart_gene_info</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;oncogene&quot;</span><span class="p">,</span> <span class="n">biomart_gene_info</span><span class="o">$</span><span class="n">description</span><span class="p">),]</span>
<span class="n">biomart_gene_info</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;tumor protein&quot;</span><span class="p">,</span> <span class="n">biomart_gene_info</span><span class="o">$</span><span class="n">description</span><span class="p">),]</span>

<span class="c1"># search for genes that have TP53 in their gene symbol:</span>
<span class="n">biomart_gene_info</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;TP53&quot;</span><span class="p">,</span> <span class="n">biomart_gene_info</span><span class="o">$</span><span class="n">external_gene_name</span><span class="p">),]</span>
</code></pre></div>
<p>For conversion of human gene symbols to mouse homologs for example (or vice versa if you provide mouse genes as the &ldquo;values&rdquo; arguments
                                                                    and &ldquo;hsapiens_homolog_associated_gene_name&rdquo; in the list of the attributes argument), you can also use biomaRt.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Convert human genes to mouse homologs</span>
<span class="n">ensembl_human_to_mouse</span> <span class="o">&lt;-</span> <span class="nf">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s">&quot;external_gene_name&quot;</span><span class="p">,</span> <span class="s">&quot;mmusculus_homolog_associated_gene_name&quot;</span><span class="p">),</span>
                                <span class="n">filter</span><span class="o">=</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">,</span>
                                <span class="n">values</span><span class="o">=</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">ensembl_gene_id</span><span class="p">,</span>
                                <span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">ensembl_human_to_mouse</span><span class="p">)</span>
<span class="c1"># ensembl_gene_id external_gene_name mmusculus_homolog_associated_gene_name</span>
<span class="c1"># 1 ENSG00000000003             TSPAN6                                 Tspan6</span>
<span class="c1"># 2 ENSG00000000419               DPM1                                   Dpm1</span>
<span class="c1"># 3 ENSG00000000419               DPM1                                Gm20716</span>
<span class="c1"># 4 ENSG00000000457              SCYL3                                  Scyl3</span>
<span class="c1"># 5 ENSG00000000460           C1orf112                               BC055324</span>
<span class="c1"># 6 ENSG00000000938                FGR                                    Fgr</span>
</code></pre></div>
<h2 id="msigdbr-package">msigdbr package</h2>
<p>The <a href="https://cran.r-project.org/web/packages/msigdbr/index.html">msigdbr</a> package hosted on CRAN allows to access gene set collections hosted on MSigDB directly within R. Check out its vignette to view how to download collections for other species such as mouse, zebra fish, horse, etc. The function <code>msigdbr_species()</code> allows you to list available species. </p>
<p>For example, to download the Hallmark collection with human gene symbols within R:
<div class="highlight"><pre><span></span><code><span class="n">gmt</span> <span class="o">&lt;-</span> <span class="n">msigdbr</span><span class="o">::</span><span class="nf">msigdbr</span><span class="p">(</span><span class="n">species</span> <span class="o">=</span> <span class="s">&quot;human&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="s">&quot;H&quot;</span><span class="p">)</span>

<span class="c1"># Create the 2 column-format (TERM2GENE argument) required by clusterProfiler:</span>
<span class="n">h_gmt</span> <span class="o">&lt;-</span> <span class="n">gmt</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs_name&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">)]</span>
</code></pre></div></p>
<p>To determine what other species are available with msigdbr:</p>
<div class="highlight"><pre><span></span><code><span class="nf">msigdbr_species</span><span class="p">()</span>
<span class="c1">#&gt; # A tibble: 20 × 2</span>
<span class="c1">#&gt;    species_name                    species_common_name                          </span>
<span class="c1">#&gt;    &lt;chr&gt;                           &lt;chr&gt;                                        </span>
<span class="c1">#&gt;  1 Anolis carolinensis             Carolina anole, green anole                  </span>
<span class="c1">#&gt;  2 Bos taurus                      bovine, cattle, cow, dairy cow, domestic cat…</span>
<span class="c1">#&gt;  3 Caenorhabditis elegans          NA                                           </span>
<span class="c1">#&gt;  4 Canis lupus familiaris          dog, dogs                                    </span>
<span class="c1">#&gt;  5 Danio rerio                     leopard danio, zebra danio, zebra fish, zebr…</span>
<span class="c1">#&gt;  6 Drosophila melanogaster         fruit fly                                    </span>
<span class="c1">#&gt;  7 Equus caballus                  domestic horse, equine, horse                </span>
<span class="c1">#&gt;  8 Felis catus                     cat, cats, domestic cat                      </span>
<span class="c1">#&gt;  9 Gallus gallus                   bantam, chicken, chickens, Gallus domesticus</span>
<span class="c1">#&gt; 10 Homo sapiens                    human                                        </span>
<span class="c1">#&gt; 11 Macaca mulatta                  rhesus macaque, rhesus macaques, Rhesus monk…</span>
<span class="c1">#&gt; 12 Monodelphis domestica           gray short-tailed opossum                    </span>
<span class="c1">#&gt; 13 Mus musculus                    house mouse, mouse                           </span>
<span class="c1">#&gt; 14 Ornithorhynchus anatinus        duck-billed platypus, duckbill platypus, pla…</span>
<span class="c1">#&gt; 15 Pan troglodytes                 chimpanzee                                   </span>
<span class="c1">#&gt; 16 Rattus norvegicus               brown rat, Norway rat, rat, rats             </span>
<span class="c1">#&gt; 17 Saccharomyces cerevisiae        baker&#39;s yeast, brewer&#39;s yeast, S. cerevisiae</span>
<span class="c1">#&gt; 18 Schizosaccharomyces pombe 972h- NA                                           </span>
<span class="c1">#&gt; 19 Sus scrofa                      pig, pigs, swine, wild boar                  </span>
<span class="c1">#&gt; 20 Xenopus tropicalis              tropical clawed frog, western clawed frog</span>
</code></pre></div>
<p>For example, we can download the Gene Ontology Biological Process (GO:BP) collection with yeast gene symbols. To first check the available collections with msigdbr, use ``msigdbr_collections()```. </p>
<div class="highlight"><pre><span></span><code><span class="nf">print</span><span class="p">(</span><span class="nf">msigdbr_collections</span><span class="p">(),</span> <span class="n">n</span><span class="o">=</span><span class="m">20</span><span class="p">)</span>
<span class="c1"># A tibble: 23 × 3</span>
<span class="c1">#    gs_cat gs_subcat         num_genesets</span>
<span class="c1">#    &lt;chr&gt;  &lt;chr&gt;                    &lt;int&gt;</span>
<span class="c1">#  1 C1     &quot;&quot;                         299</span>
<span class="c1">#  2 C2     &quot;CGP&quot;                     3384</span>
<span class="c1">#  3 C2     &quot;CP&quot;                        29</span>
<span class="c1">#  4 C2     &quot;CP:BIOCARTA&quot;              292</span>
<span class="c1">#  5 C2     &quot;CP:KEGG&quot;                  186</span>
<span class="c1">#  6 C2     &quot;CP:PID&quot;                   196</span>
<span class="c1">#  7 C2     &quot;CP:REACTOME&quot;             1615</span>
<span class="c1">#  8 C2     &quot;CP:WIKIPATHWAYS&quot;          664</span>
<span class="c1">#  9 C3     &quot;MIR:MIRDB&quot;               2377</span>
<span class="c1"># 10 C3     &quot;MIR:MIR_Legacy&quot;           221</span>
<span class="c1"># 11 C3     &quot;TFT:GTRD&quot;                 518</span>
<span class="c1"># 12 C3     &quot;TFT:TFT_Legacy&quot;           610</span>
<span class="c1"># 13 C4     &quot;CGN&quot;                      427</span>
<span class="c1"># 14 C4     &quot;CM&quot;                       431</span>
<span class="c1"># 15 C5     &quot;GO:BP&quot;                   7658</span>
<span class="c1"># 16 C5     &quot;GO:CC&quot;                   1006</span>
<span class="c1"># 17 C5     &quot;GO:MF&quot;                   1738</span>
<span class="c1"># 18 C5     &quot;HPO&quot;                     5071</span>
<span class="c1"># 19 C6     &quot;&quot;                         189</span>
<span class="c1"># 20 C7     &quot;IMMUNESIGDB&quot;             4872</span>

<span class="c1"># Obtain the C5 category and GO:BP subcategory for yeast:</span>

<span class="n">gmt</span> <span class="o">&lt;-</span> <span class="n">msigdbr</span><span class="o">::</span><span class="nf">msigdbr</span><span class="p">(</span><span class="n">species</span> <span class="o">=</span> <span class="s">&quot;Saccharomyces cerevisiae&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="s">&quot;C5&quot;</span><span class="p">,</span> <span class="n">subcategory</span> <span class="o">=</span> <span class="s">&quot;GO:BP&quot;</span><span class="p">)</span>

<span class="c1"># We obtained the GO:BP gene sets, the yeast gene symbols, and yeast ensembl_id:</span>
<span class="nf">head</span><span class="p">(</span><span class="n">gmt</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs_name&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">,</span> <span class="s">&quot;ensembl_gene&quot;</span><span class="p">)],</span> <span class="n">n</span><span class="o">=</span><span class="m">10</span><span class="p">)</span>
<span class="c1"># A tibble: 10 × 3</span>
<span class="c1"># gs_name                                          gene_symbol ensembl_gene</span>
<span class="c1"># &lt;chr&gt;                                            &lt;chr&gt;       &lt;chr&gt;       </span>
<span class="c1">#   1 GOBP_10_FORMYLTETRAHYDROFOLATE_METABOLIC_PROCESS ADE3        YGR204W     </span>
<span class="c1"># 2 GOBP_10_FORMYLTETRAHYDROFOLATE_METABOLIC_PROCESS ADE3        YGR204W     </span>
<span class="c1"># 3 GOBP_10_FORMYLTETRAHYDROFOLATE_METABOLIC_PROCESS MIS1        YBR084W     </span>
<span class="c1"># 4 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     BOL2        YGL220W     </span>
<span class="c1"># 5 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     GRX4        YER174C     </span>
<span class="c1"># 6 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     GRX5        YPL059W     </span>
<span class="c1"># 7 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     JAC1        YGL018C     </span>
<span class="c1"># 8 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     NFS1        YCL017C     </span>
<span class="c1"># 9 GOBP_2_OXOGLUTARATE_METABOLIC_PROCESS            ARO8        YGL202W     </span>
<span class="c1"># 10 GOBP_2_OXOGLUTARATE_METABOLIC_PROCESS            ADH4        YGL256W    </span>

<span class="c1"># If your DE genes are also labeled with gene symbol, create the 2 column-format (TERM2GENE argument) required by clusterProfiler for the GSEA() function for example:</span>
<span class="n">y_gmt</span> <span class="o">&lt;-</span> <span class="n">gmt</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs_name&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">)]</span>
</code></pre></div>
<p>We can also use the msigdbr package to download the KEGG collection, in case the <code>gseKEGG()</code> function of the clusterProfiler package is blocked by firewall issues.</p>
<div class="highlight"><pre><span></span><code><span class="c1">### ---- KEGG with msigdbr download:</span>
<span class="nf">msigdbr_collections</span><span class="p">()</span> <span class="c1"># </span>
<span class="c1"># C2     &quot;CP:KEGG&quot;      </span>
<span class="n">kegg_pw</span><span class="o">&lt;-</span><span class="nf">msigdbr</span><span class="p">(</span><span class="n">species</span> <span class="o">=</span> <span class="s">&quot;human&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="s">&quot;C2&quot;</span><span class="p">,</span> <span class="n">subcategory</span> <span class="o">=</span> <span class="s">&quot;CP:KEGG&quot;</span><span class="p">)</span>

<span class="nf">head</span><span class="p">(</span><span class="n">kegg_pw</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs_name&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">)])</span>
<span class="c1"># # A tibble: 6 × 2</span>
<span class="c1"># gs_name               gene_symbol</span>
<span class="c1"># &lt;chr&gt;                 &lt;chr&gt;      </span>
<span class="c1">#   1 KEGG_ABC_TRANSPORTERS ABCA1      </span>
<span class="c1"># 2 KEGG_ABC_TRANSPORTERS ABCA10     </span>
<span class="c1"># 3 KEGG_ABC_TRANSPORTERS ABCA12     </span>
<span class="c1"># 4 KEGG_ABC_TRANSPORTERS ABCA13     </span>
<span class="c1"># 5 KEGG_ABC_TRANSPORTERS ABCA2      </span>
<span class="c1"># 6 KEGG_ABC_TRANSPORTERS ABCA3      </span>

<span class="c1"># Run with GSEA function: </span>
<span class="n">kegg_gsea</span><span class="o">&lt;-</span><span class="nf">GSEA</span><span class="p">(</span><span class="n">gl</span><span class="p">,</span> <span class="n">eps</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">TERM2GENE</span> <span class="o">=</span> <span class="n">kegg_pw</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs_name&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">)],</span> <span class="n">seed</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">kegg_gsea</span><span class="o">@</span><span class="n">result</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="o">:</span><span class="m">6</span><span class="p">)])</span>
<span class="c1">#                                                                                   Description setSize enrichmentScore       NES       pvalue</span>
<span class="c1"># KEGG_RIBOSOME                                                                   KEGG_RIBOSOME      86      -0.9016245 -3.534228 9.828038e-49</span>
<span class="c1"># KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY     104       0.6183274  2.509621 7.254908e-13</span>
<span class="c1"># KEGG_REGULATION_OF_ACTIN_CYTOSKELETON                   KEGG_REGULATION_OF_ACTIN_CYTOSKELETON     170       0.4547466  1.980859 2.337476e-07</span>
<span class="c1"># KEGG_CHEMOKINE_SIGNALING_PATHWAY                             KEGG_CHEMOKINE_SIGNALING_PATHWAY     161       0.4463380  1.928426 4.354445e-07</span>
<span class="c1"># KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS                   KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS      87       0.5343886  2.100787 1.599037e-06</span>
<span class="c1"># KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY                 KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY      70       0.5299578  1.993277 8.000155e-06</span>

<span class="n">kegg_gsea</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;NATURAL_KILLER&quot;</span><span class="p">,</span> <span class="n">kegg_gsea</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="p">),</span> <span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="o">:</span><span class="m">6</span><span class="p">)]</span>
<span class="c1">#                                     Description setSize enrichmentScore      NES       pvalue</span>
<span class="c1">#  KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY     104       0.6183274 2.509621 7.254908e-13</span>
</code></pre></div>
<h2 id="code-for-barplots-with-ggplot2">Code for barplots with ggplot2</h2>
<p>Barplot of the 10 most significant gene sets:</p>
<p><div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>

<span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span> <span class="o">%&gt;%</span>
  <span class="n">dplyr</span><span class="o">::</span><span class="nf">arrange</span><span class="p">(</span><span class="n">p.adjust</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="nf">slice_head</span><span class="p">(</span><span class="n">n</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">p.adjust</span><span class="p">),</span> <span class="n">y</span> <span class="o">=</span> <span class="nf">reorder</span><span class="p">(</span><span class="n">Description</span><span class="p">,</span> <span class="o">-</span><span class="n">p.adjust</span><span class="p">)))</span> <span class="o">+</span>
    <span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span> <span class="o">=</span> <span class="s">&quot;identity&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">geom_vline</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="m">0.05</span><span class="p">),</span> <span class="n">linetype</span> <span class="o">=</span> <span class="s">&quot;dashed&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">labs</span><span class="p">(</span><span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Description&quot;</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">theme_classic</span><span class="p">()</span>
</code></pre></div>
<br />
<figure>
  <img src="../assets/images/barplot_p_value.png" width="500"/>
  </figure>
</p>
<p>Barplot of NES colored according to direction:
<div class="highlight"><pre><span></span><code><span class="n">sorted_GO_NK_Th</span> <span class="o">%&gt;%</span>
  <span class="n">dplyr</span><span class="o">::</span><span class="nf">group_by</span><span class="p">(</span><span class="n">color</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="n">dplyr</span><span class="o">::</span><span class="nf">arrange</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">NES</span><span class="p">)))</span> <span class="o">%&gt;%</span>
  <span class="nf">slice_head</span><span class="p">(</span><span class="n">n</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span> <span class="o">%&gt;%</span>
  <span class="nf">ggplot</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">NES</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="nf">reorder</span><span class="p">(</span><span class="n">Description</span><span class="p">,</span> <span class="n">NES</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">color</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span> <span class="o">=</span> <span class="s">&quot;identity&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_vline</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="m">0</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">labs</span><span class="p">(</span><span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Description&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme_classic</span><span class="p">()</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span>
</code></pre></div>
<br />
<figure>
  <img src="../assets/images/barplot_NES.png" width="700"/>
  </figure>
</p>
<h2 id="code-for-a-heatmap-of-p-values-with-ggplot2">Code for a heatmap of p-values with ggplot2</h2>
<p>For heatmaps, ggplot2 can also be used. Here is an example for a heatmap of the p-values of 6 different gene sets (gs), and the
p-value for each gene set was calculated in two comparisons, so we compared the enrichment in genes differentially expressed
between NK cells and Th cells, and between NK cells and CD8 T cells (note that this is a dummy example just to show you the 
                                                                     example of the code, it is not based on real RNA seq data).
Before using ggplot2, you need to create a dataframe that contains a column with the gene set identities, a column with the 
name of the cell type comparisons, and a column with the p-value of each gene set in each comparison. So a specific format is 
required for ggplot2.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Create a data frame that contains the p-value for every gene set for every </span>
<span class="c1"># cell type comparison. You need to include the values also for the non-significant</span>
<span class="c1"># gene sets. If you use function of the clusterProfiler package, you will usually</span>
<span class="c1"># obtain results in the @result slot only for significant gene sets. But if you need</span>
<span class="c1"># p-values also for the non-significant gene sets, you can change the argument</span>
<span class="c1"># pvalueCutoff = 1 in the gseGO function (and other functions of the clusterProfiler</span>
<span class="c1"># package)</span>

<span class="c1"># For the example, create a dummy data frame with the list of gene sets (gs), the</span>
<span class="c1"># 2 cell type comparisons, and the p-value for each gene set in each comparison:</span>
<span class="n">ora_to_plot</span><span class="o">&lt;-</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">cbind</span><span class="p">(</span><span class="n">ID</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs1&quot;</span><span class="p">,</span> <span class="s">&quot;gs2&quot;</span><span class="p">,</span><span class="s">&quot;gs3&quot;</span><span class="p">,</span><span class="s">&quot;gs4&quot;</span><span class="p">,</span> <span class="s">&quot;gs5&quot;</span><span class="p">,</span> <span class="s">&quot;gs6&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;gs1&quot;</span><span class="p">,</span> <span class="s">&quot;gs2&quot;</span><span class="p">,</span><span class="s">&quot;gs3&quot;</span><span class="p">,</span><span class="s">&quot;gs4&quot;</span><span class="p">,</span> <span class="s">&quot;gs5&quot;</span><span class="p">,</span> <span class="s">&quot;gs6&quot;</span><span class="p">),</span>
                                 <span class="n">comparison</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;NK_vs_Th&quot;</span><span class="p">,</span><span class="m">6</span><span class="p">),</span> <span class="nf">rep</span><span class="p">(</span><span class="s">&quot;NK_vs_CD8&quot;</span><span class="p">,</span><span class="m">6</span><span class="p">)),</span>
                                 <span class="n">p_val</span><span class="o">=</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0.8</span><span class="p">,</span><span class="m">0.9</span><span class="p">,</span><span class="m">0.6</span><span class="p">,</span> <span class="m">0.054</span><span class="p">,</span> <span class="m">0.00001</span><span class="p">,</span> <span class="m">0.0003</span><span class="p">,</span>
                                                    <span class="m">0.0002</span><span class="p">,</span> <span class="m">0.004</span><span class="p">,</span> <span class="m">0.001</span><span class="p">,</span><span class="m">0.01</span><span class="p">,</span> <span class="m">0.85</span><span class="p">,</span><span class="m">0.9</span><span class="p">))))</span>

<span class="c1"># transform the p-val to -log10(p-val):</span>
<span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="o">&lt;--</span><span class="nf">log10</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="o">$</span><span class="n">p_val</span><span class="p">))</span>
<span class="c1"># set the none-significant p-values to 0 so that they appear grey in the heatmap:</span>
<span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="o">&lt;-</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="o">&lt;</span><span class="p">(</span><span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="m">0.05</span><span class="p">)),</span> <span class="m">0</span><span class="p">,</span> <span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="p">)</span>
<span class="nf">range</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="o">$</span><span class="n">log10_p_val</span><span class="p">)</span> <span class="c1"># [1]  0 5</span>

<span class="nf">head</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="m">12</span><span class="p">)</span>
<span class="c1">#     ID comparison p_val log10_p_val</span>
<span class="c1"># 1  gs1   NK_vs_Th   0.8    0.000000</span>
<span class="c1"># 2  gs2   NK_vs_Th   0.9    0.000000</span>
<span class="c1"># 3  gs3   NK_vs_Th   0.6    0.000000</span>
<span class="c1"># 4  gs4   NK_vs_Th 0.054    0.000000</span>
<span class="c1"># 5  gs5   NK_vs_Th 1e-05    5.000000</span>
<span class="c1"># 6  gs6   NK_vs_Th 3e-04    3.522879</span>
<span class="c1"># 7  gs1  NK_vs_CD8 2e-04    3.698970</span>
<span class="c1"># 8  gs2  NK_vs_CD8 0.004    2.397940</span>
<span class="c1"># 9  gs3  NK_vs_CD8 0.001    3.000000</span>
<span class="c1"># 10 gs4  NK_vs_CD8  0.01    2.000000</span>
<span class="c1"># 11 gs5  NK_vs_CD8  0.85    0.000000</span>
<span class="c1"># 12 gs6  NK_vs_CD8   0.9    0.000000</span>

<span class="c1"># create a palette of colors based on the Plasma palette (grDevices package)</span>
<span class="nf">plot</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="nf">hcl.colors</span><span class="p">(</span><span class="m">20</span><span class="p">,</span><span class="s">&quot;Plasma&quot;</span><span class="p">),</span> <span class="n">pch</span><span class="o">=</span><span class="m">15</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">3</span><span class="p">)</span>
<span class="nf">hcl.colors</span><span class="p">(</span><span class="m">9</span><span class="p">,</span><span class="s">&quot;Plasma&quot;</span><span class="p">)</span><span class="c1"># &quot;RdYlGn&quot;)</span>
<span class="n">breaks</span><span class="o">&lt;-</span><span class="nf">seq</span><span class="p">(</span><span class="n">from</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">to</span><span class="o">=</span><span class="m">5</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="m">0.5</span> <span class="p">)</span>

<span class="n">color</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;grey78&quot;</span><span class="p">,</span> <span class="nf">rev</span><span class="p">(</span><span class="nf">hcl.colors</span><span class="p">(</span><span class="m">9</span><span class="p">,</span><span class="s">&quot;Plasma&quot;</span><span class="p">)))</span>

<span class="c1"># create plot and export as png:</span>
<span class="n">p</span><span class="o">&lt;-</span><span class="nf">ggplot</span><span class="p">(</span><span class="n">ora_to_plot</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">comparison</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span> <span class="n">log10_p_val</span><span class="p">))</span> <span class="o">+</span> 
  <span class="nf">geom_tile</span><span class="p">(</span><span class="n">color</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_fill_gradientn</span><span class="p">(</span><span class="n">breaks</span><span class="o">=</span> <span class="n">breaks</span><span class="p">,</span> 
                       <span class="n">colors</span> <span class="o">=</span> <span class="n">color</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme_bw</span><span class="p">()</span> 
<span class="nf">ggsave</span><span class="p">(</span><span class="n">plot</span> <span class="o">=</span> <span class="n">p</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="s">&quot;heatmap_p_value_ORA.png&quot;</span><span class="p">,</span> 
       <span class="n">device</span><span class="o">=</span><span class="s">&quot;png&quot;</span><span class="p">,</span>
       <span class="n">width</span> <span class="o">=</span> <span class="m">4</span><span class="p">,</span><span class="n">height</span> <span class="o">=</span> <span class="m">6</span><span class="p">)</span>
</code></pre></div>
<p>You will obtain the following heatmap:</p>
<figure>
  <img src="../assets/images/heatmap_p_value_ORA.png" width="300"/>
  </figure>

<h2 id="code-for-a-lollipop-plot-with-ggplot2">Code for a lollipop plot with ggplot2</h2>
<p>In this lollipop of GSEA results of leukocyte-related GO pathways, the color will represent up- (red) or down-regulated (blue) gene sets, the dot size represents the setSize. The length of the gene set description is truncated to 50 characters. We use the GSEA results that we obtained after selecting the leukocyte-related GO pathways at the end of exercise 3.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Bonus: lollipop plot of p-values of leukocyte-related GO pathways. The </span>
<span class="c1"># color will represent up- (red) or down-regulated (blue) gene sets, </span>
<span class="c1"># the dot size represents the setSize.</span>
<span class="c1"># The length of the gene set description is truncated to 50 characters:</span>

<span class="c1"># We use the GSEA results that we obtained after selecting the leukocyte-related GO pathways at the end of exercise 3</span>
<span class="n">df</span> <span class="o">&lt;-</span> <span class="n">GO_NK_Th_selection</span><span class="o">@</span><span class="n">result</span>
<span class="nf">head</span><span class="p">(</span><span class="n">df</span><span class="p">[,</span><span class="m">1</span><span class="o">:</span><span class="m">8</span><span class="p">])</span>
<span class="c1">#                    ID                       Description setSize enrichmentScore       NES       pvalue   p.adjust     qvalue</span>
<span class="c1"># GO:0097529 GO:0097529       myeloid leukocyte migration     182       0.3358543  1.459329 0.0060333962 0.04957978 0.03849908</span>
<span class="c1"># GO:0030595 GO:0030595              leukocyte chemotaxis     186       0.3314487  1.446607 0.0057064461 0.04766950 0.03701573</span>
<span class="c1"># GO:0007159 GO:0007159      leukocyte cell-cell adhesion     335       0.2955485  1.381506 0.0037782543 0.03815146 0.02962490</span>
<span class="c1"># GO:0001776 GO:0001776             leukocyte homeostasis      99      -0.3891370 -1.552210 0.0037570540 0.03812616 0.02960525</span>
<span class="c1"># GO:0002685 GO:0002685 regulation of leukocyte migration     185       0.3418586  1.489716 0.0020160142 0.02593948 0.02014220</span>
<span class="c1"># GO:0070661 GO:0070661           leukocyte proliferation     273       0.3247110  1.493837 0.0009132421 0.01545176 0.01199841</span>

<span class="n">df</span><span class="o">$</span><span class="n">mycolor</span> <span class="o">&lt;-</span> <span class="nf">ifelse</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">NES</span><span class="o">&lt;</span><span class="m">0</span><span class="p">,</span> <span class="s">&quot;cornflowerblue&quot;</span><span class="p">,</span><span class="s">&quot;indianred2&quot;</span><span class="p">)</span>

<span class="n">df</span> <span class="o">&lt;-</span> <span class="n">df</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">pvalue</span><span class="p">,</span> <span class="n">decreasing</span> <span class="o">=</span> <span class="bp">T</span><span class="p">),]</span> <span class="c1">#revert the order of the rows</span>
<span class="n">df</span><span class="o">$</span><span class="n">Label</span> <span class="o">&lt;-</span><span class="n">stringr</span><span class="o">::</span><span class="nf">str_trunc</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">Description</span><span class="p">,</span> <span class="m">50</span><span class="p">,</span> <span class="s">&quot;right&quot;</span><span class="p">)</span> <span class="c1"># keep max 50 character in a string (nchar(&quot;string&quot;) to count characters)</span>
<span class="n">df</span><span class="o">$</span><span class="n">Label</span> <span class="o">&lt;-</span> <span class="nf">factor</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">Label</span><span class="p">,</span> <span class="n">levels</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">Label</span><span class="p">))</span>

<span class="c1"># Use limits of the x-axis range according to the range of p.values:</span>
<span class="n">max_x</span> <span class="o">&lt;-</span> <span class="nf">round</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">p.adjust</span><span class="p">)),</span> <span class="n">digits</span> <span class="o">=</span> <span class="m">0</span><span class="p">)</span> <span class="o">+</span> <span class="m">0.5</span>
<span class="n">xlimits</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="n">max_x</span><span class="p">)</span>
<span class="n">xbreaks</span> <span class="o">&lt;-</span> <span class="nf">seq_along</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="n">max_x</span><span class="p">)</span>

<span class="c1"># Create the plot:</span>
<span class="n">p</span> <span class="o">&lt;-</span> <span class="nf">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">Label</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">p.adjust</span><span class="p">),</span> <span class="n">fill</span> <span class="o">=</span> <span class="n">mycolor</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">geom_segment</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">Label</span><span class="p">,</span> <span class="n">xend</span> <span class="o">=</span> <span class="n">Label</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">yend</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">p.adjust</span><span class="p">)),</span>
               <span class="n">color</span> <span class="o">=</span> <span class="n">df</span><span class="o">$</span><span class="n">mycolor</span><span class="p">,</span> <span class="n">lwd</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">pch</span> <span class="o">=</span> <span class="m">21</span><span class="p">,</span>  <span class="n">bg</span> <span class="o">=</span> <span class="n">df</span><span class="o">$</span><span class="n">mycolor</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">df</span><span class="o">$</span><span class="n">setSize</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="n">df</span><span class="o">$</span><span class="n">mycolor</span><span class="p">)</span> <span class="o">+</span> 
  <span class="nf">scale_size</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;Number\nof genes&quot;</span><span class="p">)</span> <span class="o">+</span> 
  <span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="nf">expression</span><span class="p">(</span><span class="s">&quot;-&quot;</span><span class="o">*</span><span class="s">&quot;log&quot;</span><span class="p">[</span><span class="m">10</span><span class="p">]</span><span class="o">*</span><span class="s">&quot;(p-value)&quot;</span><span class="p">),</span> 
                     <span class="n">limits</span><span class="o">=</span><span class="n">xlimits</span><span class="p">,</span>
                     <span class="n">breaks</span><span class="o">=</span><span class="n">xbreaks</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">scale_x_discrete</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme_bw</span><span class="p">(</span><span class="n">base_size</span><span class="o">=</span><span class="m">10</span><span class="p">,</span> <span class="n">base_family</span> <span class="o">=</span> <span class="s">&quot;Helvetica&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text</span><span class="o">=</span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">12</span><span class="p">,</span> <span class="n">colour</span> <span class="o">=</span> <span class="s">&quot;black&quot;</span><span class="p">),</span>
        <span class="n">axis.title</span><span class="o">=</span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">14</span><span class="p">))</span> <span class="o">+</span>
  <span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Leukocyte-related GO pathways&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">coord_flip</span><span class="p">()</span>

<span class="n">p</span>
</code></pre></div>
<p>You will obtain the following lollipop plot :</p>
<figure>
  <img src="../assets/images/lollipop.png" width="700"/>
  </figure>

<h2 id="small-information-on-a-typical-rnaseq-workflow">Small information on a typical RNAseq workflow</h2>
<p><a class="md-button" href="../assets/pdf/brief_RNAseq_workflow.pdf">Download pdf</a></p>
<h2 id="example-for-de-analysis-with-deseq2">Example for DE analysis with DESeq2</h2>
<p><a class="md-button" href="../assets/exercises/htseq_counts_NK_Th.csv">Download data</a>
<a class="md-button" href="../assets/exercises/DGE_DESeq2_example.R">Download script</a></p>
<div class="highlight"><pre><span></span><code><span class="c1"># DGE example with DESeq2:</span>
<span class="c1"># R version 4.1.2 (2021-11-01)</span>


<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;DESeq2&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span> <span class="c1"># v 1.34.0</span>

<span class="c1"># setwd(&quot;path/to/downloadedData&quot;)</span>

<span class="n">counts_NK_Th</span><span class="o">&lt;-</span><span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;htseq_counts_NK_Th.csv&quot;</span><span class="p">,</span> <span class="n">row.names</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>

<span class="n">counts_NK_Th</span><span class="o">&lt;-</span><span class="n">counts_NK_Th</span><span class="p">[</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">)</span><span class="o">==</span><span class="m">0</span><span class="p">)),]</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">)</span>
<span class="c1"># [1] 38573    15</span>

<span class="c1"># build a sample metadata table:</span>
<span class="n">coldata</span><span class="o">&lt;-</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">cbind</span><span class="p">(</span><span class="n">cell_type</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;NK&quot;</span><span class="p">,</span> <span class="m">6</span><span class="p">),</span>
                           <span class="nf">rep</span><span class="p">(</span><span class="s">&quot;Th&quot;</span><span class="p">,</span> <span class="m">9</span><span class="p">)),</span>
               <span class="n">donor</span><span class="o">=</span><span class="nf">sapply</span><span class="p">(</span><span class="nf">strsplit</span><span class="p">(</span><span class="nf">colnames</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">),</span> <span class="s">&quot;_&quot;</span><span class="p">),</span> <span class="s">&#39;[&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">),</span>
               <span class="n">sample_id</span><span class="o">=</span><span class="nf">colnames</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">)))</span>
<span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="o">&lt;-</span><span class="nf">as.factor</span><span class="p">(</span><span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="p">)</span>
<span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="o">&lt;-</span><span class="nf">factor</span><span class="p">(</span><span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="p">,</span>
                          <span class="n">levels</span><span class="o">=</span><span class="nf">levels</span><span class="p">(</span><span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="p">)[</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">)])</span>

<span class="nf">head</span><span class="p">(</span><span class="n">coldata</span><span class="p">)</span>
<span class="c1">#   cell_type donor         sample_id</span>
<span class="c1"># 1        NK   S15    S15_NK_CD56dim</span>
<span class="c1"># 2        NK   S15 S15_NK_CD56bright</span>
<span class="c1"># 3        NK   S16    S16_NK_CD56dim</span>
<span class="c1"># 4        NK   S16 S16_NK_CD56bright</span>
<span class="c1"># 5        NK   S17    S17_NK_CD56dim</span>
<span class="c1"># 6        NK   S17 S17_NK_CD56bright</span>

<span class="c1"># Create DESeq object:</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span> <span class="o">=</span> <span class="n">counts_NK_Th</span><span class="p">,</span>
                              <span class="n">colData</span> <span class="o">=</span> <span class="n">coldata</span><span class="p">,</span>
                              <span class="n">design</span><span class="o">=</span> <span class="o">~</span> <span class="n">donor</span> <span class="o">+</span> <span class="n">cell_type</span><span class="p">)</span> <span class="c1"># Difference between cell types, accounting for the sample pairing</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="nf">resultsNames</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span> <span class="c1"># lists the coefficients</span>
<span class="c1"># [1] &quot;Intercept&quot;          &quot;donor_S16_vs_S15&quot;   &quot;donor_S17_vs_S15&quot;   &quot;cell_type_NK_vs_Th&quot;</span>

<span class="n">deseq2_NK_vs_Th</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span>  
                                         <span class="n">alpha</span><span class="o">=</span><span class="m">0.05</span><span class="p">,</span> 
                                         <span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cell_type&quot;</span><span class="p">,</span><span class="s">&quot;NK&quot;</span><span class="p">,</span><span class="s">&quot;Th&quot;</span><span class="p">),</span>
                                 <span class="n">cooksCutoff</span><span class="o">=</span><span class="bp">F</span><span class="p">))</span> <span class="c1"># use cooksCutoff=F only if some genes of interest do not have a calculated p-value</span>
                                                <span class="c1"># author recomendation is to use default cooksCutoff=T    </span>

<span class="nf">head</span><span class="p">(</span><span class="n">deseq2_NK_vs_Th</span><span class="p">)</span>
<span class="c1">#            baseMean log2FoldChange     lfcSE        stat       pvalue         padj</span>
<span class="c1"># TSPAN6     40.851111     -6.9583034 1.1044621 -6.30017417 2.973114e-10 8.742152e-09</span>
<span class="c1"># TNMD        0.104281     -0.1228379 3.1336380 -0.03919977 9.687311e-01           NA</span>
<span class="c1"># DPM1     2566.964652     -0.1466129 0.2211112 -0.66307338 5.072836e-01 7.133950e-01</span>
<span class="c1"># SCYL3     571.791633      0.5728065 0.4039864  1.41788547 1.562242e-01 3.515869e-01</span>
<span class="c1"># C1orf112  201.504414      0.8758449 0.5938469  1.47486651 1.402484e-01 3.263445e-01</span>
<span class="c1"># FGR      8793.900467      8.5188295 1.2025099  7.08420757 1.398422e-12 5.868549e-11</span>

<span class="n">deseq2_NK_vs_Th</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;CPS1&quot;</span><span class="p">,</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">deseq2_NK_vs_Th</span><span class="p">)),]</span>
<span class="c1">#            baseMean log2FoldChange    lfcSE        stat     pvalue      padj</span>
<span class="c1"># CPS1     2.34916186    -3.56324252 1.855768 -1.92009033 0.05484649 0.1702518</span>
<span class="c1"># CPS1.IT1 0.09375824    -0.08381473 3.134376 -0.02674048 0.97866673        NA</span>
<span class="n">deseq2_NK_vs_Th</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;GZMB&quot;</span><span class="p">,</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">deseq2_NK_vs_Th</span><span class="p">)),]</span>
<span class="c1">#      baseMean log2FoldChange     lfcSE     stat       pvalue        padj</span>
<span class="c1"># GZMB 27758.47       9.075387 0.8897603 10.19981 1.986563e-24 2.65665e-22</span>
</code></pre></div>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        <a href="../exercises/" class="md-footer__link md-footer__link--prev" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Exercises
            </div>
          </div>
        </a>
      
      
        <a href="../links/" class="md-footer__link md-footer__link--next" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Useful links
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../assets/javascripts/workers/search.d351de03.min.js", "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.34eae1b6.min.js"></script>
      
    
  </body>
</html>