# ------------------------------------------------
# Enrichment analysis, SIB course
# ------------------------------------------------
# :8724 # R version 4.4.1 

# load the packages needed for the R exercise
library(clusterProfiler) #  v 4.12.0
library(enrichplot) # v 1.24.0
library(pathview)
library(org.Hs.eg.db) # v 3.19.1
library(ggplot2)
library(ggrepel)
library(msigdbr) # v 7.5.1

library(tidyverse) # for bonus code/dplyr/pipe

# Some reminders about the usage of R:
# to look for help on a function and which arguments to change use ?
?t.test
# to search for a keyword in all or in a particular package, use ??
??adjust
??stats::adjust

# set the working directory to where the data files are located:
setwd("/export/scratch/twyss/SIB_training/EA_062024/data/")
# set seed
set.seed(1234)

#### ---- Exercise 1

# Differential gene expression between two different 
# immune cell types, natural killer (NK) cells and CD4 T helper cells (Th)).
# The data was generated by RNA sequencing and analyzed using limma.
# The file NK_vs_Th_diff_gene_exercise_1.csv that contains gene ensembl ID,
# gene symbols, log2 fold change, t-statistic and p-values

# Steps: 
# Import csv file NK_vs_Th_diff_gene_exercise_1.csv and gmt files of gene sets.
# Fisher test of adaptive immune response for genes up-regulated in Th cells
# Hypergeometric test of 3 genesets for genes up-regulated in NK
# Import table:
NK_vs_Th<-read.csv("NK_vs_Th_diff_gene_exercise_1.csv",
                   header = T)
# Look at table structure:
head(NK_vs_Th)
# Search for a gene symbol in the data.frame, eg NCAM1 (CD56)
NK_vs_Th[which(NK_vs_Th$symbol=="NCAM1"),]

# Check the effect of p-value adjustment in data.frame:
NK_vs_Th[which(NK_vs_Th$symbol=="CPS1"),]
NK_vs_Th[which(NK_vs_Th$symbol=="GZMB"),]

# Determine whether the genes up-regulated in Th cells (i.e.down-regulated in NK cells)
# are enriched in the adaptive immune response gene set:
# http://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?geneSetName=GOBP_ADAPTIVE_IMMUNE_RESPONSE&keywords=adaptive%20immune%20response

# Import the adaptive immune response gene set (gmt file)
adaptive<-clusterProfiler::read.gmt("GOBP_ADAPTIVE_IMMUNE_RESPONSE.v7.5.1.gmt")
nrow(adaptive)
View(adaptive)
length(which(NK_vs_Th$symbol %in% adaptive$gene)) # 513

# Extract the number of genes up-regulated in Th (i.e. down-regulated in NK):
Th_up<-subset(NK_vs_Th, 
              NK_vs_Th$p.adj<=0.05&NK_vs_Th$logFC<0)
# Are these genes part of the gene set?            
summary(Th_up$symbol %in% adaptive$gene)
#    Mode   FALSE    TRUE 
# logical    1753     142 
table(Th_up$symbol %in% adaptive$gene)
length(which(Th_up$symbol %in% adaptive$gene))

## -- Number of genes up-regulated in NK (i.e. down in Th)
Th_down<-subset(NK_vs_Th, NK_vs_Th$p.adj<=0.05 & NK_vs_Th$logFC>0)
table(Th_down$symbol %in% adaptive$gene)
# FALSE  TRUE 
#. 1875    82 
length(which(Th_down$symbol %in% adaptive$gene)) # 82

# Extract the number of not significant genes:
Th_not_DE<-subset(NK_vs_Th, NK_vs_Th$p.adj>0.05)
# Are these genes part of the gene set?            
summary(Th_not_DE$symbol %in% adaptive$gene)
#    Mode   FALSE    TRUE 
# logical   16344     289

# Build contingency table
# Format:
#              up not_up
# in_set        .      .
# not_in_set    .      .

# 3: reversed rows of table outputs:
cont.table<-matrix(c(rev(unlist(table(Th_up$symbol %in% adaptive$gene))), 
                     rev(unlist(table(Th_not_DE$symbol %in% adaptive$gene)))), 
                   ncol=2, byrow = F)
cont.table

# Name the columns and rows:
colnames(cont.table)<-c("up", "not_up")
rownames(cont.table)<-c("in_set", "not_in_set")
cont.table
#              up   not_up
# in_set      142      289
# not_in_set 1753    16344
142/(142+1753)
289/(289+16344)

fisher.test(cont.table)
# Fisher's Exact Test for Count Data
# data:  cont.table
# p-value < 2.2e-16
# alternative hypothesis: true odds ratio is not equal to 1
# 95 percent confidence interval:
#  3.697701 5.654348
# sample estimates:
# odds ratio 
#   4.580549 
# 

# Test 3 gene sets among the genes up-regulated in NK cells,
# with enricher(), hypergeometric test implementation in clusterProfiler:
# Obtain the genes up-regulated in NK:

nk_up_genes<-subset(NK_vs_Th, NK_vs_Th$logFC>0 & NK_vs_Th$p.adj<=0.05)$symbol

# Help: hypergeometric test (like one-sided Fisher test = greater)
?enricher
# Import 2 other gene sets, 1 un-related to immune cells:
hair<-read.gmt("GOBP_HAIR_CELL_DIFFERENTIATION.v7.5.1.gmt")
dim(hair)
cell_active<-read.gmt("GOBP_CELL_ACTIVATION.v7.5.1.gmt")
dim(cell_active)

genesets3<-rbind(adaptive, hair, cell_active)

hyper_3genesets<-enricher(gene=nk_up_genes,
                          universe = NK_vs_Th$symbol,
                          TERM2GENE = genesets3,
                          maxGSSize = 1000)
class(hyper_3genesets)
View(hyper_3genesets@result)

# Bonus: volcano plot of DE genes:
# highlight genes of adaptive immune response in a volcano plot:
sig_genes<-subset(NK_vs_Th, NK_vs_Th$symbol %in% adaptive$gene & NK_vs_Th$p.adj<=0.05)
sig_genes_label<-subset(sig_genes, sig_genes$p.adj<=0.00001)

ggplot(NK_vs_Th, aes(x = logFC,  # 
                     y = -log10(p.adj))) +
  geom_point(color="grey87")  +
  ggtitle("Genes belonging to the adaptive immune response gene set") +
  theme_bw() +
  geom_text_repel(data = sig_genes_label, 
                  aes(x=logFC,
                      y=-log10(p.adj),label=symbol),
                  max.overlaps = 20) +
  geom_point(data=sig_genes, col="dodgerblue2") +
  theme(legend.position = "none") +
  scale_x_continuous(name = expression("log"[2]*"(fold change), NK vs Th cells")) +
  scale_y_continuous(name = expression("-"*"log"[10]*"(adj. p-value)")) +
  geom_hline(yintercept = -log10(0.05), linetype="dashed") +
  geom_vline(xintercept = 0, linetype="dashed")


## not shown Bonus: over-representation analysis of genes up-regulated in Th cells:
up_reg_th <- NK_vs_Th$symbol[NK_vs_Th$logFC < 0 & NK_vs_Th$p.adj < 0.05 ]  
hyper_3genesets_Th<-enricher(gene=up_reg_th,
                             universe = NK_vs_Th$symbol,
                             TERM2GENE = genesets3,
                             maxGSSize = 1000)
View(hyper_3genesets_Th@result)


# --------------- Exercise 2
# GSEA: prepare gene symbol-named vector of sorted t-statistics
# gsea using built-in GO genesets
idType(OrgDb = "org.Hs.eg.db") # to determine allowed gene id type

?gseGO()

gl<-NK_vs_Th$t
names(gl)<-make.names(NK_vs_Th$symbol, unique = T)
gl<-gl[order(gl, decreasing = T)]
head(gl)
tail(gl)

# already ran:
GO_NK_Th<-gseGO(gl, ont="BP",
                OrgDb = org.Hs.eg.db,
                keyType = "SYMBOL",
                minGSSize=30,
                eps=0,
                seed=T)

# To have a glance at the results table
class(GO_NK_Th)
str(GO_NK_Th)
View(GO_NK_Th@result)

# Is the adaptive immune response gene set significant?
GO_NK_Th@result[GO_NK_Th@result$Description=="adaptive immune response",]
summary(GO_NK_Th@result$p.adjust<0.05 & GO_NK_Th@result$NES<0)
# Mode       FALSE    TRUE 
# logical     320      63 
summary(GO_NK_Th@result$p.adjust<0.05 & GO_NK_Th@result$NES>0)
#     Mode   FALSE    TRUE 
# logical       63     320   

# Simplify:
?simplify()

# Already ran:
GO_NK_Th_simplify <- clusterProfiler::simplify(GO_NK_Th)
View(GO_NK_Th_simplify@result)
GO_NK_Th_simplify@result[GO_NK_Th_simplify@result$Description == "adaptive immune response",]

# you can export the results to a csv file (to summarize results using Revigo, see presentation)
write.csv(GO_NK_Th@result, "GO_GSEA_NK_vs_Th.csv",
          row.names = F, quote = F)
# save the gseaResult object:
saveRDS(GO_NK_Th, "GO_NK_Th.rds")

# How to obtain the list of leading edge genes for a gene set:
unlist(strsplit(GO_NK_Th@result[GO_NK_Th@result$Description=="adaptive immune response",11],
                "\\/"))

# How to obtain the list of all genes included in a GO term:
GO_NK_Th@geneSets$`GO:0002250`


# Bonus: Single cell RNA seq: list of significant genes tested against GO gene sets with
# over-representation analysis:

# ORA of GO:
nk_up_genes<-subset(NK_vs_Th, NK_vs_Th$logFC>0&NK_vs_Th$p.adj<=0.05)$symbol

# Already ran:
GO_enrich<-enrichGO(gene=nk_up_genes,
                    OrgDb = org.Hs.eg.db,
                    keyType = "SYMBOL",
                    # ont="BP",
                    ont="MF",
                    minGSSize = 30, 
                    universe=NK_vs_Th$symbol) # may bug with barplot below if added, depending on version

View(GO_enrich@result)
range(unlist(lapply(GO_enrich@geneSets, length)))
class(GO_enrich)
class(GO_NK_Th)

# Bonus: Histogram of the length of GO gene sets:
hist(unlist(lapply(GO_NK_Th@geneSets, length)), breaks = c(seq(0,500,by=100),seq(1000,20000,by=1000)), xlab="Length of GO gene sets", main="")

# --------------- Exercise 3: visualization

# Barplot of -log10 p-values of top 10 GO gene sets:

par(mar=c(5,15,3,3))
graphics::barplot(rev(-log10(GO_NK_Th@result$p.adjust[1:10])),
                  horiz = T, 
                  names=rev(GO_NK_Th@result$Description[1:10]),
                  las=2, xlab="-log10(adj.p-value)",
                  cex.names = 0.7,
                  col="lightgreen") 
abline(v=-log10(0.05))

# Barplot of NES of the top GO gene sets, with different color for up- or down-reg gene sets:
sorted_GO_NK_Th<- GO_NK_Th@result[order(GO_NK_Th@result$NES, decreasing = F),]
sorted_GO_NK_Th$color<-ifelse(sorted_GO_NK_Th$NES<0, "red", "darkblue")

par(mar=c(5,20,3,3))
barplot(sorted_GO_NK_Th$NES[c(1:10, (nrow(sorted_GO_NK_Th)-9):nrow(sorted_GO_NK_Th))],
        horiz = T, names=sorted_GO_NK_Th$Description[c(1:10, (nrow(sorted_GO_NK_Th)-9):nrow(sorted_GO_NK_Th))],
        las=2, xlab="NES",
        cex.names = 0.7,
        col=sorted_GO_NK_Th$color[c(1:10, (nrow(sorted_GO_NK_Th)-9):nrow(sorted_GO_NK_Th))]) 
abline(v=0)

# barplot() can be directly used on enrichResult objects: but not on gseaResult objects
graphics::barplot(GO_enrich)
graphics::barplot(GO_enrich, color = "qvalue", x = "GeneRatio")

graphics::barplot(GO_NK_Th) # this is a gseaResult object
# Error in barplot.default(GO_NK_Th) : 
# 'height' must be a vector or a matrix

# Select only 2 out of the significant gene sets: if ont="MF" was selected:
ego_selection = GO_enrich[GO_enrich@result$ID == "GO:0042287" | GO_enrich@result$ID == "GO:0004713", asis=T]
barplot(ego_selection)

# Bonus code: barplot with ggplot2
# Barplot of top 10 significant gene sets:

p1<-GO_NK_Th@result %>%
  dplyr::arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  ggplot(aes(x = -log10(p.adjust), y = reorder(Description, -p.adjust))) +
  geom_bar(stat = "identity") +
  geom_vline(xintercept = -log10(0.05), linetype = "dashed") +
  labs(y = "Description") +
  theme_classic()
p1
# ggsave(filename = "barplot_p_value.png", plot = p1, width = 5, height = 5,
#       device = "png")

p2<-sorted_GO_NK_Th %>%
  dplyr::group_by(color) %>%
  dplyr::arrange(desc(abs(NES))) %>%
  slice_head(n = 10) %>%
  ggplot(aes(x = NES, y = reorder(Description, NES), fill = color)) +
  geom_bar(stat = "identity") +
  geom_vline(xintercept = 0) +
  labs(y = "Description") +
  theme_classic() +
  theme(legend.position = "none")
p2
# ggsave(filename = "barplot_NES.png", plot = p2, width = 7, height = 5,
#        device = "png")

# dplyr::mutate(GO_enrich, qscore = -log(p.adjust, base=10)) %>% 
#   barplot(x="qscore")


# Barcode plot (i.e. gsea plot)
# You need the ID of the GO gene set to plot:
GO_NK_Th@result[1:10,1:6]

# For a gene set that is down-regulated in NK cells:
gseaplot(GO_NK_Th, geneSetID = "GO:0002181",
         title="GO:0002181 - cytoplasmic translation")
# And one that is up-regulated in NK cells
gseaplot(GO_NK_Th, geneSetID = "GO:0002443",
         title="GO:0002443 - leukocyte mediated immunity")

# Dotplot on enrichResult and gseaResult objects:
enrichplot::dotplot(GO_enrich, orderBy="p.adjust")
enrichplot::dotplot(GO_NK_Th, orderBy="p.adjust")
enrichplot::dotplot(GO_NK_Th, x="Count")

# Gene concept network:
cnetplot(GO_enrich, categorySize="pvalue") # 3 significant gene sets
cnetplot(GO_NK_Th, showCategory = 3)
cnetplot(GO_NK_Th, showCategory = 3, foldChange = gl)

# The enrichment map:
ego2 <- pairwise_termsim(GO_NK_Th)
emapplot(ego2, color="p.adjust")

# The ridge plots:
?ridgeplot
# Distribution of t-statistic for genes included in significant gene sets or in selected gene sets:
ridgeplot(GO_NK_Th)

p<-ridgeplot(GO_NK_Th, core_enrichment = F) + 
  theme(axis.text.y = element_text(size = 10)) 
ggsave("ridgeplot_EA.pdf", plot = p,
       width = 5, height = 12) 


# Select which GO terms to show in the ridge plot:
GO_NK_Th_selection <- GO_NK_Th[GO_NK_Th$ID == "GO:0002181", asis=T]
ridgeplot(GO_NK_Th_selection)
GO_NK_Th_selection <- GO_NK_Th[GO_NK_Th$ID %in% c("GO:0002181","GO:0022613",
                                                  "GO:0042254"), 
                               asis=T]
ridgeplot(GO_NK_Th_selection)
# Terms that contain the keyword "leukocyte"
GO_NK_Th_selection <- GO_NK_Th[grep("leukocyte",GO_NK_Th@result$Description), asis=T]
ridgeplot(GO_NK_Th_selection)


# Bonus: lollipop plot of p-values of leukocyte-related GO pathways. The 
# color will represent up- (red) or down-regulated (blue) gene sets, 
# the dot size represents the setSize.
# The length of the gene set description is truncated to 50 characters:
View(GO_NK_Th_selection@result)
df <- GO_NK_Th_selection@result
head(df[,1:8])
#                    ID                       Description setSize enrichmentScore       NES       pvalue   p.adjust     qvalue
# GO:0097529 GO:0097529       myeloid leukocyte migration     182       0.3358543  1.459329 0.0060333962 0.04957978 0.03849908
# GO:0030595 GO:0030595              leukocyte chemotaxis     186       0.3314487  1.446607 0.0057064461 0.04766950 0.03701573
# GO:0007159 GO:0007159      leukocyte cell-cell adhesion     335       0.2955485  1.381506 0.0037782543 0.03815146 0.02962490
# GO:0001776 GO:0001776             leukocyte homeostasis      99      -0.3891370 -1.552210 0.0037570540 0.03812616 0.02960525
# GO:0002685 GO:0002685 regulation of leukocyte migration     185       0.3418586  1.489716 0.0020160142 0.02593948 0.02014220
# GO:0070661 GO:0070661           leukocyte proliferation     273       0.3247110  1.493837 0.0009132421 0.01545176 0.01199841
df$mycolor <- ifelse(df$NES<0, "cornflowerblue","indianred2")

df <- df[order(df$pvalue, decreasing = T),] #revert the order of the rows
df$Label <-stringr::str_trunc(df$Description, 50, "right") #keep max 50 character in a string (nchar("string") to count characters) #if I run this line before cutting the 20 pathways it will give an error due to duplicated factors.
df$Label <- factor(df$Label, levels = c(df$Label))

# Use limits of the x-axis range according to the range of p.values:
max_x <- round(max(-log10(df$p.adjust)), digits = 0) + 0.5
xlimits <- c(0, max_x)
xbreaks <- seq_along(0:max_x)

# Create the plot:
p <- ggplot(df, aes(x = Label, y = -log10(p.adjust), fill = mycolor)) +
  geom_segment(aes(x = Label, xend = Label, y = 0, yend = -log10(p.adjust)),
               color = df$mycolor, lwd = 1) +
  geom_point(pch = 21,  bg = df$mycolor, aes(size=df$setSize), color=df$mycolor) + 
  scale_size(name = "Number\nof genes") + 
  scale_y_continuous(name=expression("-"*"log"[10]*"(p-value)"), 
                     limits=xlimits,
                     breaks=xbreaks) +
  scale_x_discrete(name="") +
  theme_bw(base_size=10, base_family = "Helvetica") +
  theme(axis.text=element_text(size=12, colour = "black"),
        axis.title=element_text(size=14)) +
  ggtitle("Leukocyte-related GO pathways") +
  coord_flip()

p

#### ---- Exercise 4

# - First convert the gene symbols to EntrezID to perform a GSEA of KEGG 
# pathways (with argument minGSSize=30). 
# - Are the majority of gene sets rather up-regulated or down-regulated?
# - Is there a KEGG immune-related gene set coming up? Is there a KEGG Natural 
# killer gene set coming up?
# - If you want to see which genes are included in one of the built-in KEGG pathways, 
# where could you find this information?
# - Import the hallmark gene sets and run a GSEA. How many significant gene sets are there?

# Check allowed gene ID types:
?gseKEGG
keytypes(org.Hs.eg.db)
# convert from= "ENSEMBL" to "SYMBOL" and "ENTREZID"
gene_convert <- bitr(as.character(NK_vs_Th$ensembl_gene_id), 
                     fromType="ENSEMBL", 
                     toType=c("SYMBOL", "ENTREZID"), OrgDb="org.Hs.eg.db",
                     drop=TRUE)
head(gene_convert)
dim(gene_convert)

# GSEA of KEGG: create a vector of genes that are coded with the EntrezID:
# use the sorted gene list gl previously created:
# Create a vector of genes that are coded with the EntrezID:
# use the sorted gene list gl previously created:
gl_kegg<-cbind(SYMBOL=names(gl), t=gl)

# merge with converted gene symbols to combine both:
# by default the data frames are merged on the columns with names they both have
gl_kegg<-merge(gl_kegg, gene_convert)
head(gl_kegg)

gl_kegg_list<-as.numeric(as.character(gl_kegg$t))
names(gl_kegg_list)<-as.character(gl_kegg$ENTREZID)
gl_kegg_list<-sort(gl_kegg_list, decreasing = T)

# run GSEA of KEGG:
# Already ran:
KEGG_NK_Th<-gseKEGG(gl_kegg_list, 
                    organism = "hsa", 
                    keyType = "ncbi-geneid",
                    minGSSize = 30,
                    eps=0,
                    seed=T)
# Reading KEGG annotation online: "https://rest.kegg.jp/link/hsa/pathway"...
# Reading KEGG annotation online: "https://rest.kegg.jp/list/pathway/hsa"...
# Reading KEGG annotation online: "https://rest.kegg.jp/conv/ncbi-geneid/hsa"...

# Explore the object:
class(KEGG_NK_Th)
str(KEGG_NK_Th)
View(KEGG_NK_Th@result)

# Up-regulated gene sets:
summary(KEGG_NK_Th@result$NES>0)
#    Mode   FALSE    TRUE 
# logical      7      21 

KEGG_NK_Th@result[grep("immune",KEGG_NK_Th@result$Description), ]

KEGG_NK_Th@result[grep("Natural killer",KEGG_NK_Th@result$Description), ]
#                ID                               Description setSize enrichmentScore     NES       pvalue     p.adjust
# hsa04650 hsa04650 Natural killer cell mediated cytotoxicity      97        0.614284 2.49564 5.483815e-12 4.825757e-10
# qvalue rank                   leading_edge
# hsa04650 3.867532e-10 1873 tags=52%, list=13%, signal=45%

# How many built-in KEGG pathways are there?
length(KEGG_NK_Th@geneSets)
# Genes involved in Natural killer cell mediated cytotoxicity (hsa04650)
KEGG_NK_Th@geneSets$hsa04650 # coded as Entrez Gene ID...

# Overlay of fold change value on KEGG pathway figure:
library(pathview)

# pathview map with non-significant genes in grey:
# set log fold change of non-significant genes to 0:
NK_vs_Th$logFC_0<-ifelse(NK_vs_Th$p.adj>0.05, 0, NK_vs_Th$logFC)

# create named vector of fold change values:
genePW<-NK_vs_Th$logFC_0
names(genePW)<-NK_vs_Th$symbol

# Create pathview map for Ribosome = hsa03010
pathview(gene.data  = genePW,
         pathway.id = "hsa03010",
         species    = "hsa",
         gene.idtype = "SYMBOL")

# Create pathview map of Natural killer cell mediated cytotoxicity = hsa04650
pathview(gene.data  = genePW,
         pathway.id = "hsa04650",
         species    = "hsa",
         gene.idtype = "SYMBOL")

# Import hallmark, convert to term2gene and run GSEA:
term2gene_h <- msigdbr(species = "Homo sapiens", category = "H")
# Or alternatively:
# term2gene_h<-read.gmt("h.all.v2023.2.Hs.symbols.gmt")

head(term2gene_h)
length(unique(term2gene_h$gs_name)) # 50

# Run GSEA with the function that allows to use custom gene sets, 
# provide the named vector of t statistics
h_NK_vs_Th<-GSEA(gl, 
                 TERM2GENE = term2gene_h[,c("gs_name", "gene_symbol")],
                 eps=0,
                 seed=T)

View(h_NK_vs_Th@result)

# Number of significant gene sets:
length(which(h_NK_vs_Th@result$p.adjust<=0.05)) 

# A quick dotplot of the hallmark results:
dotplot(h_NK_vs_Th)
dotplot(h_NK_vs_Th, x="NES", orderBy="p.adjust")

# A barcode plot:
gseaplot2(h_NK_vs_Th, geneSetID = "HALLMARK_MTORC1_SIGNALING",
          title="HALLMARK_MTORC1_SIGNALING")


### ---- KEGG with msigdbr download:
msigdbr_collections() # 
# C2     "CP:KEGG"      
kegg_pw<-msigdbr(species = "human", category = "C2", subcategory = "CP:KEGG")

head(kegg_pw[,c("gs_name", "gene_symbol")])
# # A tibble: 6 × 2
# gs_name               gene_symbol
# <chr>                 <chr>      
#   1 KEGG_ABC_TRANSPORTERS ABCA1      
# 2 KEGG_ABC_TRANSPORTERS ABCA10     
# 3 KEGG_ABC_TRANSPORTERS ABCA12     
# 4 KEGG_ABC_TRANSPORTERS ABCA13     
# 5 KEGG_ABC_TRANSPORTERS ABCA2      
# 6 KEGG_ABC_TRANSPORTERS ABCA3      

# Run with GSEA function: 
kegg_gsea<-GSEA(gl, eps = 0, TERM2GENE = kegg_pw[,c("gs_name", "gene_symbol")], seed = TRUE)
head(kegg_gsea@result[,c(2:6)])
#                                                                                   Description setSize enrichmentScore       NES       pvalue
# KEGG_RIBOSOME                                                                   KEGG_RIBOSOME      86      -0.9016245 -3.534228 9.828038e-49
# KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY     104       0.6183274  2.509621 7.254908e-13
# KEGG_REGULATION_OF_ACTIN_CYTOSKELETON                   KEGG_REGULATION_OF_ACTIN_CYTOSKELETON     170       0.4547466  1.980859 2.337476e-07
# KEGG_CHEMOKINE_SIGNALING_PATHWAY                             KEGG_CHEMOKINE_SIGNALING_PATHWAY     161       0.4463380  1.928426 4.354445e-07
# KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS                   KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS      87       0.5343886  2.100787 1.599037e-06
# KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY                 KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY      70       0.5299578  1.993277 8.000155e-06

kegg_gsea@result[grep("NATURAL_KILLER", kegg_gsea@result$Description), c(2:6,11)]
#                                     Description setSize enrichmentScore      NES       pvalue
#  KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY     104       0.6183274 2.509621 7.254908e-13


### --- Bonus code: msigdbr C5 GO:BP collection for yeast:
gmt <- msigdbr::msigdbr(species = "Saccharomyces cerevisiae", category = "C5", subcategory = "GO:BP")

# We obtained the GO:BP gene sets, the yeast gene symbols, and yeast ensembl_id:
head(gmt[,c("gs_name", "gene_symbol", "ensembl_gene")], n=10)
# A tibble: 10 × 3
# gs_name                                          gene_symbol ensembl_gene
# <chr>                                            <chr>       <chr>       
#   1 GOBP_10_FORMYLTETRAHYDROFOLATE_METABOLIC_PROCESS ADE3        YGR204W     
# 2 GOBP_10_FORMYLTETRAHYDROFOLATE_METABOLIC_PROCESS ADE3        YGR204W     
# 3 GOBP_10_FORMYLTETRAHYDROFOLATE_METABOLIC_PROCESS MIS1        YBR084W     
# 4 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     BOL2        YGL220W     
# 5 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     GRX4        YER174C     
# 6 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     GRX5        YPL059W     
# 7 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     JAC1        YGL018C     
# 8 GOBP_2FE_2S_CLUSTER_ASSEMBLY                     NFS1        YCL017C     
# 9 GOBP_2_OXOGLUTARATE_METABOLIC_PROCESS            ARO8        YGL202W     
# 10 GOBP_2_OXOGLUTARATE_METABOLIC_PROCESS            ADH4        YGL256W    

# If your DE genes are also labeled with gene symbol, create the 2 column-format (TERM2GENE argument) required by clusterProfiler for the GSEA() function for example:
y_gmt <- gmt[,c("gs_name", "gene_symbol")]

###### --- Bonus code: Conversion with biomaRt
BiocManager::install("biomaRt")
library(biomaRt)
listEnsembl()

ensembl <- useEnsembl(biomart = "genes")

datasets <- listDatasets(ensembl)
head(datasets)
# search for human dataset and genome version:
datasets[grep("sapiens", datasets$dataset),]
# 80 hsapiens_gene_ensembl Human genes (GRCh38.p13) GRCh38.p13

# For human:
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")

# you can extract gene biotype, description, wiki gene description,
# chromosome location of genes, even strand information etc
attributes <- listAttributes(ensembl)
attributes[1:15,]

biomart_gene_info <- getBM(attributes=c("ensembl_gene_id", "external_gene_name", "description", 
                                        "gene_biotype", "wikigene_description"),
                           filter="ensembl_gene_id",
                           values=NK_vs_Th$ensembl_gene_id,
                           mart=ensembl)
head(biomart_gene_info)
biomart_gene_info[grep("oncogene", biomart_gene_info$description),]
biomart_gene_info[grep("tumor protein", biomart_gene_info$description),]

biomart_gene_info[grep("TP53", biomart_gene_info$external_gene_name),]

# Convert human genes to mouse homologs
ensembl_human_to_mouse <- getBM(attributes=c("ensembl_gene_id","external_gene_name", "mmusculus_homolog_associated_gene_name"),
                                filter="ensembl_gene_id",
                                values=NK_vs_Th$ensembl_gene_id,
                                mart=ensembl)
head(ensembl_human_to_mouse)
# ensembl_gene_id external_gene_name mmusculus_homolog_associated_gene_name
# 1 ENSG00000000003             TSPAN6                                 Tspan6
# 2 ENSG00000000419               DPM1                                   Dpm1
# 3 ENSG00000000419               DPM1                                Gm20716
# 4 ENSG00000000457              SCYL3                                  Scyl3
# 5 ENSG00000000460           C1orf112                               BC055324
# 6 ENSG00000000938                FGR                                    Fgr


### --- Bonus code: for p-value heatmap using ggplot2:

# Create a data frame that contains the p-value for every gene set for every 
# cell type comparison. You need to include the values also for the non-significant
# gene sets. If you use function of the clusterProfiler package, you will usually
# obtain results in the @result slot only for significant gene sets. But if you need
# p-values also for the non-significant gene sets, you can change the argument
# pvalueCutoff = 1 in the gseGO function (and other functions of the clusterProfiler
# package)

# For the example, create a dummy data frame with the list of gene sets (gs), the
# 2 cell type comparisons, and the p-value for each gene set in each comparison:
ora_to_plot<-as.data.frame(cbind(ID=c("gs1", "gs2","gs3","gs4", "gs5", "gs6",
                                      "gs1", "gs2","gs3","gs4", "gs5", "gs6"),
                                 comparison=c(rep("NK_vs_Th",6), rep("NK_vs_CD8",6)),
                                 p_val=as.numeric(c(0.8,0.9,0.6, 0.054, 0.00001, 0.0003,
                                                    0.0002, 0.004, 0.001,0.01, 0.85,0.9))))
# transform the p-val to -log10(p-val):
ora_to_plot$log10_p_val<--log10(as.numeric(ora_to_plot$p_val))
# set the none-significant p-values to 0 so that they appear grey in the heatmap:
ora_to_plot$log10_p_val<-ifelse(ora_to_plot$log10_p_val<(-log10(0.05)), 0, ora_to_plot$log10_p_val)
range(ora_to_plot$log10_p_val) # [1]  0 5

head(ora_to_plot, n=12)
#     ID comparison p_val log10_p_val
# 1  gs1   NK_vs_Th   0.8    0.000000
# 2  gs2   NK_vs_Th   0.9    0.000000
# 3  gs3   NK_vs_Th   0.6    0.000000
# 4  gs4   NK_vs_Th 0.054    0.000000
# 5  gs5   NK_vs_Th 1e-05    5.000000
# 6  gs6   NK_vs_Th 3e-04    3.522879
# 7  gs1  NK_vs_CD8 2e-04    3.698970
# 8  gs2  NK_vs_CD8 0.004    2.397940
# 9  gs3  NK_vs_CD8 0.001    3.000000
# 10 gs4  NK_vs_CD8  0.01    2.000000
# 11 gs5  NK_vs_CD8  0.85    0.000000
# 12 gs6  NK_vs_CD8   0.9    0.000000
# create a palette of colors based on the Plasma palette (grDevices package)
plot(1:20, 1:20, col=hcl.colors(20,"Plasma"), pch=15, cex=3)
hcl.colors(9,"Plasma")# "RdYlGn")
breaks<-seq(from=0, to=5, by=0.5 )

color<-c("grey78", rev(hcl.colors(9,"Plasma")))

p<-ggplot(ora_to_plot, aes(comparison, ID, fill= log10_p_val)) + 
  geom_tile(color = "black") +
  scale_fill_gradientn(breaks= breaks, 
                       colors = color) +
  theme_bw() 
pdf("heatmap_p_value_ORA.pdf", width = 4, height = 6)
p
dev.off()

ggsave(plot = p, filename = "heatmap_p_value_ORA.png", 
       device="png", width = 3.5, height = 5)



#### ---- Extra exercise for credits
# - Perform GSEA of the NK vs Th data using the Reactome gene sets downloaded on the 
# MSigDB website.
# - How many gene sets are significantly enriched? Generate an ordered barplot of 
# the NES of all significant gene sets, and generate a barcode plot for the gene set with the lowest NES

# Steps
# - import the reactome gene sets in the file c2.cp.reactome.v7.1.symbols.gmt using read.gmt()
# Use GSEA providing a sorted named vector of t-statistics and the reactome gene sets, 
# use minGSSize=30
# NOTE: if the GSEA fails, use readRDS to import the pre-computed results contained in the 
# file "reactomeGSEA_NK_vs_Th_results.rds"
# - count the number of significant adjusted p-values
# - use barplot() and gseaplot() for the visualization of the results

