
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../bonus_code/">
      
      
      
      <link rel="icon" href="../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.27">
    
    
      
        <title>Useful links - Enrichment analysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.6543a935.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tools-for-rnaseq-de-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Enrichment analysis" class="md-header__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Enrichment analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Useful links
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/enrichment-analysis-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Enrichment analysis" class="md-nav__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Enrichment analysis
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/enrichment-analysis-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../precourse/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Precourse preparations
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../course_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Course schedule
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../materials/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Materials
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../exercises/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Exercises
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../bonus_code/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Bonus code
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Useful links
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Useful links
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tools-for-rnaseq-de-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Tools for RNAseq DE analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-tools-for-enrichment-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Additional tools for enrichment analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#considerations-and-challenges-of-enrichment-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Considerations and challenges of enrichment analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tools-for-species-other-than-human-or-mouse" class="md-nav__link">
    <span class="md-ellipsis">
      Tools for species other than human or mouse
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#proteomics" class="md-nav__link">
    <span class="md-ellipsis">
      Proteomics
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tools-for-rnaseq-de-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Tools for RNAseq DE analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-tools-for-enrichment-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Additional tools for enrichment analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#considerations-and-challenges-of-enrichment-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Considerations and challenges of enrichment analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tools-for-species-other-than-human-or-mouse" class="md-nav__link">
    <span class="md-ellipsis">
      Tools for species other than human or mouse
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#proteomics" class="md-nav__link">
    <span class="md-ellipsis">
      Proteomics
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Useful links</h1>

<h2 id="tools-for-rnaseq-de-analysis">Tools for RNAseq DE analysis</h2>
<p><strong>limma</strong> for differential gene expression analysis of microarray or RNA seq data, also includes functions for enrichment analysis</p>
<p><a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf">https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf</a></p>
<p><strong>edgeR</strong> for differential gene expression analysis of RNA seq data</p>
<p><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf</a></p>
<p><strong>DESeq2</strong> for differential gene expression analysis of RNA seq data</p>
<p><a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></p>
<p><strong>From reads to genes to pathways</strong> Pipeline for DGE analysis and pathway analysis</p>
<p><a href="https://f1000research.com/articles/5-1438">https://f1000research.com/articles/5-1438</a> </p>
<p><strong>bioconductor</strong>: Introduction and structure</p>
<p><a href="https://ivanek.github.io/analysisOfGenomicsDataWithR/02_IntroToBioc_html.html">https://ivanek.github.io/analysisOfGenomicsDataWithR/02_IntroToBioc_html.html</a></p>
<h2 id="additional-tools-for-enrichment-analysis">Additional tools for enrichment analysis</h2>
<p><strong>clusterProfiler</strong> e-Book</p>
<p><a href="https://yulab-smu.github.io/clusterProfiler-book/">https://yulab-smu.github.io/clusterProfiler-book/</a></p>
<p><strong>fgsea</strong> package with the <code>fgsea()</code> function called by clusterProfiler</p>
<p><a href="https://bioconductor.org/packages/release/bioc/html/fgsea.html">https://bioconductor.org/packages/release/bioc/html/fgsea.html</a></p>
<p><strong>GOsummaries</strong> bioconductor package, to visualise Gene Ontology (GO) enrichment analysis results on gene lists arising from different analyses such clustering or PCA. The significant GO categories are visualised as word clouds that can be combined with different plots summarising the underlying data.</p>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/GOsummaries.html">https://www.bioconductor.org/packages/release/bioc/html/GOsummaries.html</a></p>
<p><strong>online tool</strong> for overrepresentation analysis</p>
<p><a href="http://www.pantherdb.org/">http://www.pantherdb.org/</a></p>
<p><strong>online tool</strong> for gene set enrichment analysis</p>
<p><a href="http://www.webgestalt.org/">http://www.webgestalt.org/</a></p>
<p><strong>Revigo and rrvgo</strong></p>
<p><a href="http://revigo.irb.hr/">http://revigo.irb.hr/</a></p>
<p><a href="http://bioconductor.org/packages/release/bioc/html/rrvgo.html">http://bioconductor.org/packages/release/bioc/html/rrvgo.html</a></p>
<p>A <a href="https://academic.oup.com/nar/article/49/D1/D325/6027811">recent publication</a> of the <strong>Gene Ontology</strong> consortium. They try to improve taxon constraints, i.e. to better define
which terms are relevant for some species but not for others (such as leukocyte-related terms that should be constrained to vertebrates.</p>
<p><a href="https://www.ebi.ac.uk/chebi/init.do"><strong>ChEBI</strong></a>: A freely available dictionary of molecular entities focused on ‘small’ chemical compounds. Can be used as a database for GSEA in metabolomics studies. </p>
<h2 id="considerations-and-challenges-of-enrichment-analysis">Considerations and challenges of enrichment analysis</h2>
<p>A <a href="https://www.frontiersin.org/articles/10.3389/fgene.2020.00654/full">recent publication (2020)</a> on gene set analysis challenges, opportunities and future research. </p>
<p><strong>Importance of the ranking metrics in GSEA</strong>: Paper on the <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1674-0">&ldquo;Ranking metrics in gene set enrichment analysis: do they matter?&rdquo;</a></p>
<p><a href="https://academic.oup.com/bib/article/23/3/bbac143/6572658"><strong>Mubeen et al, 2022</strong></a>: <em>On the influence of several factors on pathway enrichment analysis</em></p>
<p><a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009935"><strong>Wijesooriya et al, 2022</strong></a>: <em>Urgent need for consistent standards in functional enrichment analysis</em></p>
<h2 id="tools-for-species-other-than-human-or-mouse">Tools for species other than human or mouse</h2>
<p>** OrgDB** packages: Bioconductor contains &ldquo;org.db&rdquo; packages with genome/gene annotation for many organisms, for example
for <a href="https://bioconductor.org/packages/release/data/annotation/html/org.Dm.eg.db.html">Drosophila</a> or <a href="https://bioconductor.org/packages/release/data/annotation/html/org.At.tair.db.html">Arabidopsis</a>.</p>
<p>For <strong>fungi</strong>, <a href="https://fungidb.org/fungidb/app/search/pathway/PathwaysByGeneList">FungiDB</a> provides identification of metabolic pathways based on a-user selected list of genes</p>
<p><strong>PlantGSAD</strong>, a gene set database for plants:  <a href="https://academic.oup.com/nar/article/50/D1/D1456/6371972?login=true">https://academic.oup.com/nar/article/50/D1/D1456/6371972?login=true</a></p>
<p>Eventually the AllEnricher tool by <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-020-3408-y">Zhang et al</a>: <em>AllEnricher: a comprehensive gene set function enrichment tool for both model and non-model species</em> </p>
<p>A <a href="https://bio-protocol.org/en/bpdetail?id=3912&amp;type=0"><em>pipeline for non-model species</em></a> using Scots pine as an example: The authors describe how to assemble the transcriptome from RNA sequencing data, how to annotate genes, and suggest GO enrichment analysis. </p>
<p><a href="https://academic.oup.com/nar/article/47/D1/D309/5173662"><strong>eggNOGG</strong></a>: <em>A hierarchical, functionally and phylogenetically annotated orthology resource based on 5090 organisms and 2502 viruses</em>. Describes a public database of orthology relationships, gene evolutionary histories and functional annotations.</p>
<!-- [**A blog post**](https://archetypalecology.wordpress.com/2021/01/27/how-to-perform-kegg-and-go-enrichment-analysis-of-non-model-species-using-r/) on performing KEGG and GO enrichment analysis on non-model species, after eggNOGG. -->

<h2 id="proteomics">Proteomics</h2>
<p><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S16-S13"><strong>Bessarabova et al, 2012</strong></a>: <em>Knowledge-based analysis of proteomics data</em>. Not the most recent paper but a nice overview. </p>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/34236662/"><strong>Combes et al, 2021</strong></a>: <em>GO Enrichment Analysis for Differential Proteomics Using ProteoRE</em>. </p>
<p><strong>Proteomics data analysis in R</strong>: <a href="https://pnnl-comp-mass-spec.github.io/proteomics-data-analysis-tutorial/">https://pnnl-comp-mass-spec.github.io/proteomics-data-analysis-tutorial/</a>. Chapter 8 contains code for ORA and GSEA with clusterProfiler, but also with fgsea. </p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
    
  </body>
</html>