
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.1.6">
    
    
      
        <title>Exercises - Enrichment analysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.875de78c.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.f1a3b89f.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    <script>function __prefix(e){return new URL("..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#source-of-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Enrichment analysis" class="md-header__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Enrichment analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Exercises
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
      <div class="md-header__source">
        
<a href="https://github.com/sib-swiss/enrichment-analysis-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Enrichment analysis" class="md-nav__button md-logo" aria-label="Enrichment analysis" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Enrichment analysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sib-swiss/enrichment-analysis-training/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/enrichment-analysis-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../materials/" class="md-nav__link">
        Materials
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Exercises
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Exercises
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#source-of-data" class="md-nav__link">
    Source of data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-1-over-representation-analysis" class="md-nav__link">
    Exercise 1 - Over-representation analysis
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-2-gene-set-enrichment-analysis-gsea" class="md-nav__link">
    Exercise 2 - Gene set enrichment analysis (GSEA)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-3-visualization-of-enrichment-results" class="md-nav__link">
    Exercise 3 - Visualization of enrichment results
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-4-the-last-one-enrichment-of-other-collections-of-gene-sets" class="md-nav__link">
    Exercise 4 (the last one!) - Enrichment of other collections of gene sets
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#feedback" class="md-nav__link">
    Feedback
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#extra-exercise-for-ects-credits" class="md-nav__link">
    Extra exercise for ECTS credits
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../bonus_code/" class="md-nav__link">
        Bonus code
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../links/" class="md-nav__link">
        Useful links
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#source-of-data" class="md-nav__link">
    Source of data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-1-over-representation-analysis" class="md-nav__link">
    Exercise 1 - Over-representation analysis
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-2-gene-set-enrichment-analysis-gsea" class="md-nav__link">
    Exercise 2 - Gene set enrichment analysis (GSEA)
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-3-visualization-of-enrichment-results" class="md-nav__link">
    Exercise 3 - Visualization of enrichment results
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercise-4-the-last-one-enrichment-of-other-collections-of-gene-sets" class="md-nav__link">
    Exercise 4 (the last one!) - Enrichment of other collections of gene sets
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#feedback" class="md-nav__link">
    Feedback
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#extra-exercise-for-ects-credits" class="md-nav__link">
    Extra exercise for ECTS credits
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/sib-swiss/enrichment-analysis-training/edit/master/docs/exercises.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Exercises</h1>
                
                <p>In this section, you will find the R code that we will use during the course. We will explain the code and output during correction of the exercises.</p>
<h2 id="source-of-data">Source of data</h2>
<p>We will work with RNA sequencing data generated by <a href="https://jlb.onlinelibrary.wiley.com/doi/10.1002/JLB.5MA0120-209R">Ercolano et al 2020</a>. This study described the transcriptomes of immune cells that are circulating in the blood of humans in healthy conditions. Different types of immune cells circulate in human blood. In this study, 2 cell types were included: Natural Killer (NK) cells and CD4+ T helper (Th) cells. These 2 types of cells have different functions: NK cells provide a rapid response in the innate immune response at the beginning of an infection such as by a virus, and are able to kill virus-infected cells by secreting cytotoxic molecules; Th cells have an important role in the adaptive immune response by aiding antigen-presenting cells by secreting different panels of cytokines depending on the type of pathogen infecting the host. Here we perform enrichment analysis using the genes that are differentially expressed between NK cells and Th cells.</p>
<p>The RNA sequencing data that we provide includes the results of a differential gene expression analysis comparing NK cells to Th cells. The differential gene expression analysis was performed using limma (see <a href="https://sib-swiss.github.io/enrichment-analysis-training/links/">Useful links</a> page). </p>
<p>Before starting the exercises, set the working directory to where you have downloaded and unzipped the <a href="https://sib-swiss.github.io/enrichment-analysis-training/materials/">data folder</a> with the files for the exercises 
and load the necessary packages. Also, set the seed so that results are replicable.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>When using <code>setwd()</code>, change the path within quotes to where you have saved the data for the exercises</p>
</div>
<div class="highlight"><pre><span></span><code><span class="c1"># Change the path here:</span>
<span class="nf">setwd</span><span class="p">(</span><span class="s">&quot;/path/to/whereDataFolderIsSaved/&quot;</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">clusterProfiler</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">enrichplot</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">pathview</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">msigdbr</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span> <span class="c1"># for bonus code/dplyr/pipe</span>


<span class="c1"># set seed</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">1234</span><span class="p">)</span>
</code></pre></div>
<h2 id="exercise-1-over-representation-analysis">Exercise 1 - Over-representation analysis</h2>
<p>Import the data into your R session and explore its structure: what are the different columns corresponding to?
How can you search for a particular gene within this table?</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Import DE table:</span>
<span class="n">NK_vs_Th</span><span class="o">&lt;-</span><span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;data/NK_vs_Th_diff_gene_exercise_1.csv&quot;</span><span class="p">,</span>
                   <span class="n">header</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>

<span class="c1"># Look at the structure of the data.frame:</span>
<span class="nf">head</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="p">)</span>

<span class="c1"># Search for a gene symbol in the data.frame, eg NCAM1 (CD56)</span>
<span class="n">NK_vs_Th</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span><span class="o">==</span><span class="s">&quot;NCAM1&quot;</span><span class="p">),]</span>
</code></pre></div>
<p>This table contains the typical information obtained after differential gene expression analysis. For each gene, you have log<sub>2</sub> fold change values and significance information via the raw p-value and adjusted p-value. A positive log<sub>2</sub> fold change value indicates that the gene is more expressed in NK cells than in Th cells, while a negative log<sub>2</sub> fold change value indicates that the gene is less expressed in NK cells than in Th cells.
The p-value is usually computed automatically by R packages designed for differential gene expression analysis such as edgeR, limma or DESeq2. If you ever need to perform p-value adjustment, there is a function that is part of the stats package called p.adjust(), that allows you to perform p-value
adjustment for any list of p-values. Look at the help of the function and try to understand the arguments that can be used. 
<div class="highlight"><pre><span></span><code><span class="o">?</span><span class="n">p.adjust</span>
</code></pre></div></p>
<p>We can compare the raw p-value and adjusted p-value for any gene. Raw p-values that are close to 0.05 will often become higher than 0.05 after adjustment,
while very small p-values more likely remain below 0.05 after adjustment. Search for 2 genes in the data.frame, CPS1 and GZMB, and verify the effect of adjustment on their p-values. Are both genes still significant after adjustment?</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="n">NK_vs_Th</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span><span class="o">==</span><span class="s">&quot;CPS1&quot;</span><span class="p">),]</span>

<span class="n">NK_vs_Th</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span><span class="o">==</span><span class="s">&quot;GZMB&quot;</span><span class="p">),]</span> 
</code></pre></div>
</details>
<p>If we expect a particular group (or set) of genes to be altered between 2 conditions, we can specifically test for this using a single Fisher test, i.e an over-representation analysis.
Here, because we compared cells that are involved in the innate immune response (NK cells) to cells involved in the adaptive immune response (Th cells), we
expect that there are differences in genes involved in the adaptive immune response. Which direction do you think the adaptive immune response genes should have? Rather up-regulated in NK or down-regulated in NK vs Th cells?</p>
<p>As you will see later in the course, there are databases available where you can obtain gene sets. We have obtained the Gene Ontology adaptive immune response gene set from the <a href="http://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?geneSetName=GOBP_ADAPTIVE_IMMUNE_RESPONSE&amp;keywords=adaptive%20immune%20response">MSigDB</a> resource.
Import the list of genes included in this gene set, and determine using a Fisher test whether the genes up-regulated in Th cells (i.e. down-regulated in NK cells) are enriched in the adaptive immune response. </p>
<div class="highlight"><pre><span></span><code><span class="c1"># Import the adaptive immune response gene set (gmt file)</span>
<span class="n">adaptive</span><span class="o">&lt;-</span><span class="n">clusterProfiler</span><span class="o">::</span><span class="nf">read.gmt</span><span class="p">(</span><span class="s">&quot;data/GOBP_ADAPTIVE_IMMUNE_RESPONSE.v7.5.1.gmt&quot;</span><span class="p">)</span>
<span class="nf">nrow</span><span class="p">(</span><span class="n">adaptive</span><span class="p">)</span> <span class="c1"># 719</span>
<span class="nf">View</span><span class="p">(</span><span class="n">adaptive</span><span class="p">)</span>
</code></pre></div>
<p>RNA sequencing analysis often involves filtering genes to remove lowly expressed genes. Therefore, it is possible that not all genes that are part of a gene set are 
found in your RNA seq dataset. Check how many genes in the RNA-seq data set are found in the adaptive immune response gene set.</p>
<div class="highlight"><pre><span></span><code><span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">))</span> <span class="c1"># 513</span>
</code></pre></div>
<p>To perform a Fisher&rsquo;s exact test, we need to build a contingency table of the number of genes that are up-regulated or not significant in Th cells, and that are part or not of the gene set of interest. First, count the number of genes up-regulated in Th cells (i.e. down-regulated in NK cells), then the number of not significant genes, then determine for each type of genes whether they are part of the adaptive immune response gene set or not.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="c1"># Extract the number of genes up-regulated in Th (i.e. down-regulated in NK):</span>
<span class="n">Th_up</span><span class="o">&lt;-</span><span class="nf">subset</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="p">,</span> 
<span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">p.adj</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="o">&amp;</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">logFC</span><span class="o">&lt;</span><span class="m">0</span><span class="p">)</span>
<span class="c1"># Are these genes part of the gene set?            </span>
<span class="nf">summary</span><span class="p">(</span><span class="n">Th_up</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">)</span>
<span class="c1">#    Mode   FALSE    TRUE </span>
<span class="c1"># logical    1753     142 </span>
<span class="nf">table</span><span class="p">(</span><span class="n">Th_up</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">)</span>

<span class="c1"># Extract the number of not significant genes:</span>
<span class="n">Th_not_DE</span><span class="o">&lt;-</span><span class="nf">subset</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="p">,</span> <span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">p.adj</span><span class="o">&gt;</span><span class="m">0.05</span><span class="p">)</span>
<span class="c1"># Are these genes part of the gene set?            </span>
<span class="nf">summary</span><span class="p">(</span><span class="n">Th_not_DE</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">)</span>
<span class="c1">#    Mode   FALSE    TRUE </span>
<span class="c1"># logical   16344     289</span>
</code></pre></div>
</details>
<p>Next, build a contingency table that has the following format:</p>
<table>
<thead>
<tr>
<th></th>
<th>Up-reg.</th>
<th>Not up-reg.</th>
</tr>
</thead>
<tbody>
<tr>
<td>In gene set</td>
<td>#</td>
<td>#</td>
</tr>
<tr>
<td>Not in gene set</td>
<td>#</td>
<td>#</td>
</tr>
</tbody>
</table>
<p>Finally, use the <code>fisher.test()</code> function to determine whether the adaptive immune response is over-represented among the genes up-regulated in Th cells. If you don&rsquo;t know how to use the `fisher.test() function, remember to use the help:</p>
<div class="highlight"><pre><span></span><code><span class="o">?</span><span class="nf">fisher.test</span><span class="p">()</span>
</code></pre></div>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="c1"># There are several ways to build the contigency table:</span>
<span class="c1"># 1: by hand:</span>
<span class="n">cont.table</span><span class="o">&lt;-</span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">142</span><span class="p">,</span> <span class="m">1753</span><span class="p">,</span> <span class="m">289</span><span class="p">,</span> <span class="m">16344</span><span class="p">),</span> <span class="n">ncol</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="n">byrow</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">cont.table</span>

<span class="c1"># 2: by using the individual values of the table() function:</span>
<span class="n">cont.table</span><span class="o">&lt;-</span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">unlist</span><span class="p">(</span><span class="nf">table</span><span class="p">(</span><span class="n">Th_up</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">))[[</span><span class="m">2</span><span class="p">]],</span>
<span class="nf">unlist</span><span class="p">(</span><span class="nf">table</span><span class="p">(</span><span class="n">Th_up</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">))[[</span><span class="m">1</span><span class="p">]],</span> 
<span class="nf">unlist</span><span class="p">(</span><span class="nf">table</span><span class="p">(</span><span class="n">Th_not_DE</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">))[[</span><span class="m">2</span><span class="p">]],</span> 
<span class="nf">unlist</span><span class="p">(</span><span class="nf">table</span><span class="p">(</span><span class="n">Th_not_DE</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">))[[</span><span class="m">1</span><span class="p">]]),</span> 
<span class="n">ncol</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">byrow</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">cont.table</span>
<span class="c1"># 3: by using the 2 rows (reversed) of the table() function:</span>
<span class="n">cont.table</span><span class="o">&lt;-</span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">rev</span><span class="p">(</span><span class="nf">unlist</span><span class="p">(</span><span class="nf">table</span><span class="p">(</span><span class="n">Th_up</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">))),</span> 
<span class="nf">rev</span><span class="p">(</span><span class="nf">unlist</span><span class="p">(</span><span class="nf">table</span><span class="p">(</span><span class="n">Th_not_DE</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span><span class="p">)))),</span> 
<span class="n">ncol</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">byrow</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
<span class="n">cont.table</span>

<span class="c1"># Add row and colum names so that you know what each number corresponds to:</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">cont.table</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;up&quot;</span><span class="p">,</span> <span class="s">&quot;not_up&quot;</span><span class="p">)</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">cont.table</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;in_set&quot;</span><span class="p">,</span> <span class="s">&quot;not_in_set&quot;</span><span class="p">)</span>
<span class="n">cont.table</span>
<span class="c1">#              up   not_up</span>
<span class="c1"># in_set      142      289</span>
<span class="c1"># not_in_set 1753    16344</span>

<span class="c1"># Compare the proportions: </span>
<span class="m">142</span><span class="o">/</span><span class="p">(</span><span class="m">142+1753</span><span class="p">)</span>
<span class="m">289</span><span class="o">/</span><span class="p">(</span><span class="m">289+16344</span><span class="p">)</span>

<span class="c1"># Run fisher test:</span>
<span class="nf">fisher.test</span><span class="p">(</span><span class="n">cont.table</span><span class="p">)</span>
<span class="c1"># Fisher&#39;s Exact Test for Count Data</span>
<span class="c1"># data:  cont.table</span>
<span class="c1"># p-value &lt; 2.2e-16</span>
<span class="c1"># alternative hypothesis: true odds ratio is not equal to 1</span>
<span class="c1"># 95 percent confidence interval:</span>
<span class="c1">#  3.697701 5.654348</span>
<span class="c1"># sample estimates:</span>
<span class="c1"># odds ratio </span>
<span class="c1">#   4.580549 </span>
</code></pre></div>
</details>
<p>In case you wish to run an over-representation analysis for several gene sets at once, instead of running several independent Fisher tests, you can use the <code>enricher()</code> function of the clusterProfiler package. This function allows you to perform an over-representation analysis of several gene sets at once, using an implementation of the hypergeometric test (a one-sided Fisher&rsquo;s exact test).
Check out the help of the function: What arguments do you need to provide? </p>
<p>Determine whether the genes up-regulated in NK cells are over-represented in 3 gene sets: the adaptive immune response gene set used above, another related to cell activation, and one un-related to immune cells (hair cell differentiation).</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Help: hypergeometric test (like one-sided Fisher test = greater)</span>
<span class="o">?</span><span class="nf">enricher</span><span class="p">()</span>

<span class="c1"># Test 3 gene sets among the genes up-regulated in NK cells,</span>
<span class="c1"># with enricher()</span>
<span class="c1"># First, obtain the genes up-regulated in NK:</span>

<span class="n">nk_up_genes</span><span class="o">&lt;-</span><span class="nf">subset</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="p">,</span> <span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">logFC</span><span class="o">&gt;</span><span class="m">0</span><span class="o">&amp;</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">p.adj</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">)</span><span class="o">$</span><span class="n">symbol</span>

<span class="c1"># Import 2 other gene sets, 1 un-related to immune cells:</span>
<span class="n">hair</span><span class="o">&lt;-</span><span class="nf">read.gmt</span><span class="p">(</span><span class="s">&quot;data/GOBP_HAIR_CELL_DIFFERENTIATION.v7.5.1.gmt&quot;</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">hair</span><span class="p">)</span>
<span class="n">cell_active</span><span class="o">&lt;-</span><span class="nf">read.gmt</span><span class="p">(</span><span class="s">&quot;data/GOBP_CELL_ACTIVATION.v7.5.1.gmt&quot;</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">cell_active</span><span class="p">)</span>

<span class="c1"># Combine the 3 gene sets into a single data.frame for the TERM2GENE argument:</span>
<span class="n">genesets3</span><span class="o">&lt;-</span><span class="nf">rbind</span><span class="p">(</span><span class="n">adaptive</span><span class="p">,</span> <span class="n">hair</span><span class="p">,</span> <span class="n">cell_active</span><span class="p">)</span>

<span class="n">hyper_3genesets</span><span class="o">&lt;-</span><span class="nf">enricher</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="n">nk_up_genes</span><span class="p">,</span>
                      <span class="n">universe</span> <span class="o">=</span> <span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span><span class="p">,</span>
                      <span class="n">TERM2GENE</span> <span class="o">=</span> <span class="n">genesets3</span><span class="p">,</span>
                      <span class="n">maxGSSize</span> <span class="o">=</span> <span class="m">1000</span><span class="p">)</span>
<span class="nf">View</span><span class="p">(</span><span class="n">hyper_3genesets</span><span class="o">@</span><span class="n">result</span><span class="p">)</span>
</code></pre></div>
<p><strong>Bonus</strong> How would you represent the over-representation result for the adaptive immune response? An option might be a volcano plot coloring genes that are part of the gene set 
and significant using <a href="https://ggplot2.tidyverse.org/">ggplot2</a>.</p>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span>

<span class="n">sig_genes</span><span class="o">&lt;-</span><span class="nf">subset</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="p">,</span> <span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span> <span class="o">%in%</span> <span class="n">adaptive</span><span class="o">$</span><span class="n">gene</span> <span class="o">&amp;</span> <span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">p.adj</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">)</span>
<span class="n">sig_genes_label</span><span class="o">&lt;-</span><span class="nf">subset</span><span class="p">(</span><span class="n">sig_genes</span><span class="p">,</span> <span class="n">sig_genes</span><span class="o">$</span><span class="n">p.adj</span><span class="o">&lt;=</span><span class="m">0.00001</span><span class="p">)</span>

<span class="nf">ggplot</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">logFC</span><span class="p">,</span>  <span class="c1"># </span>
                     <span class="n">y</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">p.adj</span><span class="p">)))</span> <span class="o">+</span>
        <span class="nf">geom_point</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;grey87&quot;</span><span class="p">)</span>  <span class="o">+</span>
        <span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Genes belonging to the adaptive immune response gene set&quot;</span><span class="p">)</span> <span class="o">+</span>
        <span class="nf">theme_bw</span><span class="p">()</span> <span class="o">+</span>
        <span class="nf">geom_text_repel</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">sig_genes_label</span><span class="p">,</span> 
                        <span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">logFC</span><span class="p">,</span>
                            <span class="n">y</span><span class="o">=-</span><span class="nf">log10</span><span class="p">(</span><span class="n">p.adj</span><span class="p">),</span><span class="n">label</span><span class="o">=</span><span class="n">symbol</span><span class="p">),</span>
                        <span class="n">max.overlaps</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span> <span class="o">+</span>
        <span class="nf">geom_point</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">sig_genes</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;dodgerblue2&quot;</span><span class="p">)</span> <span class="o">+</span>
        <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span> <span class="o">+</span>
        <span class="nf">scale_x_continuous</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="nf">expression</span><span class="p">(</span><span class="s">&quot;log&quot;</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="o">*</span><span class="s">&quot;(fold change), NK vs Th cells&quot;</span><span class="p">))</span> <span class="o">+</span>
        <span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="nf">expression</span><span class="p">(</span><span class="s">&quot;-&quot;</span><span class="o">*</span><span class="s">&quot;log&quot;</span><span class="p">[</span><span class="m">10</span><span class="p">]</span><span class="o">*</span><span class="s">&quot;(adj. p-value)&quot;</span><span class="p">))</span> <span class="o">+</span>
        <span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="m">0.05</span><span class="p">),</span> <span class="n">linetype</span><span class="o">=</span><span class="s">&quot;dashed&quot;</span><span class="p">)</span> <span class="o">+</span>
        <span class="nf">geom_vline</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">linetype</span><span class="o">=</span><span class="s">&quot;dashed&quot;</span><span class="p">)</span>
</code></pre></div>
<figure>
  <img src="../assets/images/bonus_volcano.png" width="700"/>
  </figure>

<h2 id="exercise-2-gene-set-enrichment-analysis-gsea">Exercise 2 - Gene set enrichment analysis (GSEA)</h2>
<p>Next, we will use functions of the clusterProfiler package to perform a gene set enrichment analysis (GSEA) of Gene Ontology terms. For this, we first need
to create a vector with the t-statistic of each gene, assign the gene symbol to each t-statistic in the vector (i.e. names), then sort the vector.</p>
<div class="admonition warning">
<p class="admonition-title">Warning<p>A single gene can be labeled with different types of gene labels: Ensembl IDs, NCBI Entrez IDs, gene symbols, UniProt IDs, etc. 
Therefore, first check whether the way the genes are labeled is supported by clusterProfiler using:
<div class="highlight"><pre><span></span><code><span class="nf">idType</span><span class="p">(</span><span class="n">OrgDb</span> <span class="o">=</span> <span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">)</span> <span class="c1"># to determine allowed gene id type</span>
</code></pre></div></p>
</p>
</div>
<p>Create the named and sorted vector of t-statistics (i.e. for the geneList argument of the gseGO() function). </p>
<div class="highlight"><pre><span></span><code><span class="n">gl</span><span class="o">&lt;-</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">t</span>
<span class="nf">names</span><span class="p">(</span><span class="n">gl</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">make.names</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span><span class="p">,</span> <span class="n">unique</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>
<span class="n">gl</span><span class="o">&lt;-</span><span class="n">gl</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">gl</span><span class="p">,</span> <span class="n">decreasing</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)]</span>

<span class="n">GO_NK_Th</span><span class="o">&lt;-</span><span class="nf">gseGO</span><span class="p">(</span><span class="n">gl</span><span class="p">,</span> <span class="n">ont</span><span class="o">=</span><span class="s">&quot;BP&quot;</span><span class="p">,</span>
                <span class="n">OrgDb</span> <span class="o">=</span> <span class="n">org.Hs.eg.db</span><span class="p">,</span>
                <span class="n">keyType</span> <span class="o">=</span> <span class="s">&quot;SYMBOL&quot;</span><span class="p">,</span>
                <span class="n">minGSSize</span><span class="o">=</span><span class="m">30</span><span class="p">,</span>
                <span class="n">eps</span><span class="o">=</span><span class="m">0</span><span class="p">,</span>
                <span class="n">seed</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>
</code></pre></div>
<p>Explore the new object that was created. What is its structure? What does it contain? Is the adaptive immune response gene set significant? How many gene sets are down- or up-regulated?</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code>    <span class="nf">class</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">)</span>
    <span class="nf">str</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">)</span>
    <span class="nf">View</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">)</span>

    <span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="o">==</span><span class="s">&quot;adaptive immune response&quot;</span><span class="p">,]</span>
    <span class="nf">summary</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">p.adjust</span><span class="o">&lt;</span><span class="m">0.05</span><span class="o">&amp;</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">NES</span><span class="o">&lt;</span><span class="m">0</span><span class="p">)</span>
  <span class="c1"># Mode     FALSE    TRUE </span>
  <span class="c1"># logical    290      61</span>

  <span class="nf">summary</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">p.adjust</span><span class="o">&lt;</span><span class="m">0.05</span><span class="o">&amp;</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">NES</span><span class="o">&gt;</span><span class="m">0</span><span class="p">)</span>
  <span class="c1">#     Mode   FALSE    TRUE </span>
  <span class="c1"># logical       61     290 </span>
</code></pre></div>
</details>
<p>Among the results, some gene sets seem to be a bit redundant based on the genes they contain (column &ldquo;core_enrichment&rdquo;). We can simplify the results by using the <code>simplify()</code> function of the clusterProfiler package:</p>
<p><div class="highlight"><pre><span></span><code><span class="c1"># Simplify:</span>
<span class="o">?</span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">simplify</span>
<span class="n">GO_NK_Th_simplify</span> <span class="o">&lt;-</span> <span class="n">clusterProfiler</span><span class="o">::</span><span class="nf">simplify</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">)</span>
<span class="nf">View</span><span class="p">(</span><span class="n">GO_NK_Th_simplify</span><span class="o">@</span><span class="n">result</span><span class="p">)</span>
<span class="n">GO_NK_Th_simplify</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="n">GO_NK_Th_simplify</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="o">==</span><span class="s">&quot;adaptive immune response&quot;</span><span class="p">,]</span>
</code></pre></div>
Finally, you can export the results to a csv file, if you would want to summarize the results using <a href="http://revigo.irb.hr/">Revigo</a> for example (see later in course).
To obtain the list of leading edge genes, split the corresponding column in the <code>@result</code> by using the slash. And if you want to obtain a full list
of genes belonging to a particular GO term, this information is also stored in the <code>@geneSets</code> slot.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Export results</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">,</span> <span class="s">&quot;GO_GSEA_NK_vs_Th.csv&quot;</span><span class="p">,</span>
          <span class="n">row.names</span> <span class="o">=</span> <span class="bp">F</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>

<span class="c1"># How to obtain the list of leading edge genes for a gene set:</span>
<span class="nf">unlist</span><span class="p">(</span><span class="nf">strsplit</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="o">==</span><span class="s">&quot;adaptive immune response&quot;</span><span class="p">,</span><span class="m">11</span><span class="p">],</span>
                <span class="s">&quot;\\/&quot;</span><span class="p">))</span>

<span class="c1"># How to obtain the list of all genes included in a GO term:</span>
<span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">geneSets</span><span class="o">$</span><span class="n">`GO:0002250`</span>
</code></pre></div>
<p><strong>Bonus</strong> In the context of single-cell RNA sequencing data for example, single-cells are clustered, and marker genes per cluster are obtained. In this case, we
don&rsquo;t always obtain a t-statistic or fold change value for every gene in the dataset, but rather a simple list of significant genes. With this list,
you can perform an over-representation analysis of the GO terms using enrichGO(). The output will not contain enrichment scores, but rather p-values for
every gene set.</p>
<div class="highlight"><pre><span></span><code><span class="n">nk_up_genes</span><span class="o">&lt;-</span><span class="nf">subset</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="p">,</span> <span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">logFC</span><span class="o">&gt;</span><span class="m">0</span><span class="o">&amp;</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">p.adj</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">)</span><span class="o">$</span><span class="n">symbol</span>

<span class="c1"># Provide list of significant genes for over-representation analysis of GO gene sets </span>
<span class="c1"># using enrichGO():</span>
<span class="n">GO_enrich</span><span class="o">&lt;-</span><span class="nf">enrichGO</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="n">nk_up_genes</span><span class="p">,</span>
                    <span class="n">OrgDb</span> <span class="o">=</span> <span class="n">org.Hs.eg.db</span><span class="p">,</span>
                    <span class="n">keyType</span> <span class="o">=</span> <span class="s">&quot;SYMBOL&quot;</span><span class="p">,</span>
                    <span class="n">maxGSSize</span> <span class="o">=</span> <span class="m">30</span><span class="p">)</span> <span class="c1">#</span>
                    <span class="n">universe</span><span class="o">=</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span><span class="p">)</span> <span class="c1"># may bug with barplot below if universe is added, depending on version</span>
<span class="nf">View</span><span class="p">(</span><span class="n">GO_enrich</span><span class="o">@</span><span class="n">result</span><span class="p">)</span>

<span class="c1"># Check the class of the objects created:</span>
<span class="nf">class</span><span class="p">(</span><span class="n">GO_enrich</span><span class="p">)</span>
<span class="c1"># [1] &quot;enrichResult&quot;</span>
<span class="c1"># attr(,&quot;package&quot;)</span>
<span class="c1"># [1] &quot;DOSE&quot;</span>
<span class="nf">class</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">)</span>
<span class="c1"># [1] &quot;gseaResult&quot;</span>
<span class="c1"># attr(,&quot;package&quot;)</span>
<span class="c1"># [1] &quot;DOSE&quot;</span>
</code></pre></div>
<p>This method can also be used if you have a list of genes involved in a disease, such as genes that are mutated in cancer, or lists of genes obtained by other omics methods, such as genes that have a transcription factor binding site in their promoter (ChIP-seq), etc.</p>
<h2 id="exercise-3-visualization-of-enrichment-results">Exercise 3 - Visualization of enrichment results</h2>
<p>Once we have performed a GSEA or over-representation analysis, we need of course to show the results and prepare figures for a publication. Several visualizations are possible.
The first option is a barplot. It can be a barplot of p-values of top over-represented GO terms. P-values are usually -log<sub>10</sub>-transformed, so that the very small p-values are emphasized compared to p-values that are closer to 0.05. The barplot() function comes from the graphics package. </p>
<div class="highlight"><pre><span></span><code><span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="m">20</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">3</span><span class="p">))</span>
<span class="nf">barplot</span><span class="p">(</span><span class="nf">rev</span><span class="p">(</span><span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">p.adjust</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">])),</span>
        <span class="n">horiz</span> <span class="o">=</span> <span class="bp">T</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="nf">rev</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">]),</span>
        <span class="n">las</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;-log10(adj.p-value)&quot;</span><span class="p">,</span>
        <span class="n">cex.names</span> <span class="o">=</span> <span class="m">0.7</span><span class="p">,</span>
        <span class="n">col</span><span class="o">=</span><span class="s">&quot;lightgreen&quot;</span><span class="p">)</span> 
<span class="nf">abline</span><span class="p">(</span><span class="n">v</span><span class="o">=-</span><span class="nf">log10</span><span class="p">(</span><span class="m">0.05</span><span class="p">))</span>
</code></pre></div>
<p>In publications, we often see barplots of normalized enrichment scores (NES), showing which gene sets are up-regulated and which are down-regulated.</p>
<p>Now that you know how to use the barplot() function, how would you create a barplot of NES of the top 10 up-regulated and the top 10 down-regulated genes sets,
 with red bars for the up-regulated gene sets and blue bars for the down-regulated gene sets?</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code>    <span class="c1"># Barplot of NES of the top GO gene sets, with different color for up- or down-reg gene sets:</span>
    <span class="n">sorted_GO_NK_Th</span><span class="o">&lt;-</span> <span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">NES</span><span class="p">,</span> <span class="n">decreasing</span> <span class="o">=</span> <span class="bp">F</span><span class="p">),]</span>
    <span class="n">sorted_GO_NK_Th</span><span class="o">$</span><span class="n">color</span><span class="o">&lt;-</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="o">$</span><span class="n">NES</span><span class="o">&lt;</span><span class="m">0</span><span class="p">,</span> <span class="s">&quot;red&quot;</span><span class="p">,</span> <span class="s">&quot;darkblue&quot;</span><span class="p">)</span>

    <span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="m">20</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">3</span><span class="p">))</span>
    <span class="nf">barplot</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="o">$</span><span class="n">NES</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="p">)</span><span class="m">-9</span><span class="p">)</span><span class="o">:</span><span class="nf">nrow</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="p">))],</span>
        <span class="n">horiz</span> <span class="o">=</span> <span class="bp">T</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">sorted_GO_NK_Th</span><span class="o">$</span><span class="n">Description</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="p">)</span><span class="m">-9</span><span class="p">)</span><span class="o">:</span><span class="nf">nrow</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="p">))],</span>
        <span class="n">las</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;NES&quot;</span><span class="p">,</span>
        <span class="n">cex.names</span> <span class="o">=</span> <span class="m">0.7</span><span class="p">,</span>
        <span class="n">col</span><span class="o">=</span><span class="n">sorted_GO_NK_Th</span><span class="o">$</span><span class="n">color</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="p">)</span><span class="m">-9</span><span class="p">)</span><span class="o">:</span><span class="nf">nrow</span><span class="p">(</span><span class="n">sorted_GO_NK_Th</span><span class="p">))])</span> 
    <span class="nf">abline</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="m">0</span><span class="p">)</span>
</code></pre></div>
</details>
<p>With results of an over-representation analysis, where the output is of class &ldquo;enrichResult&rdquo;, the barplot() function can be used directly. You can either show the significant gene sets, or a custom selection of them.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Use the GO_enrich analysis performed above, of the over-representation analysis </span>
<span class="c1"># of genes up-regulated in NK cells:</span>
<span class="c1"># barplot() can be directly used on enrichResult objects: but not on gseaResult objects</span>
<span class="n">graphics</span><span class="o">::</span><span class="nf">barplot</span><span class="p">(</span><span class="n">GO_enrich</span><span class="p">)</span>
<span class="n">graphics</span><span class="o">::</span><span class="nf">barplot</span><span class="p">(</span><span class="n">GO_enrich</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;qvalue&quot;</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s">&quot;GeneRatio&quot;</span><span class="p">)</span>

<span class="n">graphics</span><span class="o">::</span><span class="nf">barplot</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">)</span> <span class="c1"># this is a gseaResult object</span>
<span class="c1"># Error in barplot.default(GO_NK_Th) : </span>
<span class="c1"># &#39;height&#39; must be a vector or a matrix</span>

<span class="c1"># Select only 2 out of the significant gene sets:</span>
<span class="n">ego_selection</span> <span class="o">=</span> <span class="n">GO_enrich</span><span class="p">[</span><span class="n">GO_enrich</span><span class="o">$</span><span class="n">ID</span> <span class="o">==</span> <span class="s">&quot;GO:0019864&quot;</span> <span class="o">|</span> <span class="n">GO_enrich</span><span class="o">$</span><span class="n">ID</span> <span class="o">==</span> <span class="s">&quot;GO:0045159&quot;</span><span class="p">,</span> <span class="n">asis</span><span class="o">=</span><span class="bp">T</span><span class="p">]</span>
<span class="n">graphics</span><span class="o">::</span><span class="nf">barplot</span><span class="p">(</span><span class="n">ego_selection</span><span class="p">)</span>
</code></pre></div>
<p>Here, we used the barplot function of the graphics package, but you can of course use ggplot2 functions. Please see the <a href="https://sib-swiss.github.io/enrichment-analysis-training/bonus_code/">Bonus code</a> page for examples of barplots with ggplot2 (and tidyverse) functions. </p>
<p>A common way to represent the enrichment score of a single gene set after GSEA is the barcode plot (i.e gsea plot), with the <code>gseaplot()</code> function that can be used directly on objects of class &ldquo;gseaResult&rdquo; (i.e. output by clusterProfiler functions). </p>
<div class="highlight"><pre><span></span><code><span class="c1"># Barcode plot</span>
<span class="c1"># You need the ID of the GO gene set to plot:</span>
<span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">]</span>

<span class="c1"># For a gene set that is down-regulated in NK cells:</span>
<span class="nf">gseaplot</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">,</span> <span class="n">geneSetID</span> <span class="o">=</span> <span class="s">&quot;GO:0002181&quot;</span><span class="p">,</span>
         <span class="n">title</span><span class="o">=</span><span class="s">&quot;GO:0002181 - cytoplasmic translation&quot;</span><span class="p">)</span>
<span class="c1"># And one that is up-regulated in NK cells</span>
<span class="nf">gseaplot</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">,</span> <span class="n">geneSetID</span> <span class="o">=</span> <span class="s">&quot;GO:0002443&quot;</span><span class="p">,</span>
         <span class="n">title</span><span class="o">=</span><span class="s">&quot;GO:0002443 - leukocyte mediated immunity&quot;</span><span class="p">)</span>
</code></pre></div>
<figure>
  <img src="../assets/images/gseaplot1.png" width="700"/>
  </figure>

<figure>
  <img src="../assets/images/gseaplot2.png" width="700"/>
  </figure>

<p>Finally, clusterProfiler provides several visualization methods that can be used directly on objects of class &ldquo;gseaResult&rdquo; or &ldquo;enrichResult. These functions are implemented in the enrichplot package that was developped to go with clusterProfiler. Examples are the dotplot, the gene-concept network (cnetplot), the enrichment map (emapplot) and the ridge plots.
Feel free to consult the <a href="http://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html">chapter on visualization of the nice clusterProfiler book</a> to see all the possible options and test others than the ones listed here!</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Dotplot on enrichResult and gseaResult objects:</span>
<span class="n">enrichplot</span><span class="o">::</span><span class="nf">dotplot</span><span class="p">(</span><span class="n">GO_enrich</span><span class="p">,</span> <span class="n">orderBy</span><span class="o">=</span><span class="s">&quot;p.adjust&quot;</span><span class="p">)</span>
<span class="n">enrichplot</span><span class="o">::</span><span class="nf">dotplot</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">,</span> <span class="n">orderBy</span><span class="o">=</span><span class="s">&quot;p.adjust&quot;</span><span class="p">)</span>

<span class="c1"># Gene concept network:</span>
<span class="nf">cnetplot</span><span class="p">(</span><span class="n">GO_enrich</span><span class="p">,</span> <span class="n">categorySize</span><span class="o">=</span><span class="s">&quot;pvalue&quot;</span><span class="p">)</span>
<span class="nf">cnetplot</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">,</span> <span class="n">showCategory</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span>

<span class="c1"># The enrichment map:</span>
<span class="n">ego2</span> <span class="o">&lt;-</span> <span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">)</span>
<span class="nf">emapplot</span><span class="p">(</span><span class="n">ego2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;p.adjust&quot;</span><span class="p">)</span>

<span class="c1"># The ridge plots:</span>
<span class="c1"># Distribution of t-statistic for genes included in significant gene sets or in selected gene sets:</span>
<span class="nf">ridgeplot</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">)</span>
<span class="c1"># What is the difference with core_enrichment =F?</span>
<span class="nf">ridgeplot</span><span class="p">(</span><span class="n">GO_NK_Th</span><span class="p">,</span> <span class="n">core_enrichment</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>

<span class="c1"># Select which GO terms to show in the ridge plot:</span>
<span class="n">GO_NK_Th_selection</span> <span class="o">&lt;-</span> <span class="n">GO_NK_Th</span><span class="p">[</span><span class="n">GO_NK_Th</span><span class="o">$</span><span class="n">ID</span> <span class="o">==</span> <span class="s">&quot;GO:0002181&quot;</span><span class="p">,</span> <span class="n">asis</span><span class="o">=</span><span class="bp">T</span><span class="p">]</span>
<span class="nf">ridgeplot</span><span class="p">(</span><span class="n">GO_NK_Th_selection</span><span class="p">)</span>
<span class="n">GO_NK_Th_selection</span> <span class="o">&lt;-</span> <span class="n">GO_NK_Th</span><span class="p">[</span><span class="n">GO_NK_Th</span><span class="o">$</span><span class="n">ID</span> <span class="o">%in%</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;GO:0002181&quot;</span><span class="p">,</span><span class="s">&quot;GO:0022613&quot;</span><span class="p">,</span>
                                                  <span class="s">&quot;GO:0042254&quot;</span><span class="p">),</span> 
                                                  <span class="n">asis</span><span class="o">=</span><span class="bp">T</span><span class="p">]</span>
<span class="nf">ridgeplot</span><span class="p">(</span><span class="n">GO_NK_Th_selection</span><span class="p">)</span>
<span class="c1"># Terms that contain the keyword &quot;leukocyte&quot;</span>
<span class="n">GO_NK_Th_selection</span> <span class="o">&lt;-</span> <span class="n">GO_NK_Th</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;leukocyte&quot;</span><span class="p">,</span><span class="n">GO_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="p">),</span> <span class="n">asis</span><span class="o">=</span><span class="bp">T</span><span class="p">]</span>
<span class="nf">ridgeplot</span><span class="p">(</span><span class="n">GO_NK_Th_selection</span><span class="p">)</span>
</code></pre></div>
<h2 id="exercise-4-the-last-one-enrichment-of-other-collections-of-gene-sets">Exercise 4 (the last one!) - Enrichment of other collections of gene sets</h2>
<p>We will now perform GSEA of other collections of gene sets, such as KEGG or Hallmark.
For GSEA of the KEGG gene sets, the gene IDs have to be NCBI Entrez gene IDs. ClusterProfiler provides the bitr() function to convert gene label types.</p>
<div class="highlight"><pre><span></span><code><span class="nf">keytypes</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
<span class="c1"># convert from= &quot;ENSEMBL&quot; to &quot;SYMBOL&quot; and &quot;ENTREZID&quot;</span>
<span class="n">gene_convert</span> <span class="o">&lt;-</span> <span class="nf">bitr</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">ensembl_gene_id</span><span class="p">),</span> 
                     <span class="n">fromType</span><span class="o">=</span><span class="s">&quot;ENSEMBL&quot;</span><span class="p">,</span> 
                     <span class="n">toType</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;SYMBOL&quot;</span><span class="p">,</span> <span class="s">&quot;ENTREZID&quot;</span><span class="p">),</span> <span class="n">OrgDb</span><span class="o">=</span><span class="s">&quot;org.Hs.eg.db&quot;</span><span class="p">)</span>

<span class="c1"># Check the format of the data frame obtained after conversion:                     </span>
<span class="nf">head</span><span class="p">(</span><span class="n">gene_convert</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">gene_convert</span><span class="p">)</span>
</code></pre></div>
<p>Similarly to the gseGO() function, we need to provide a named and sorted vector of t-statistics, the difference is that the names will be Entrez gene IDs.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Create a vector of genes that are coded with the EntrezID:</span>
<span class="c1"># use the sorted gene list gl previously created:</span>
<span class="n">gl_kegg</span><span class="o">&lt;-</span><span class="nf">cbind</span><span class="p">(</span><span class="n">SYMBOL</span><span class="o">=</span><span class="nf">names</span><span class="p">(</span><span class="n">gl</span><span class="p">),</span> <span class="n">t</span><span class="o">=</span><span class="n">gl</span><span class="p">)</span>

<span class="c1"># merge with converted gene symbols to combine both:</span>
<span class="c1"># by default the data frames are merged on the columns with names they both have</span>
<span class="n">gl_kegg</span><span class="o">&lt;-</span><span class="nf">merge</span><span class="p">(</span><span class="n">gl_kegg</span><span class="p">,</span> <span class="n">gene_convert</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">gl_kegg</span><span class="p">)</span>

<span class="n">gl_kegg_list</span><span class="o">&lt;-</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">gl_kegg</span><span class="o">$</span><span class="n">t</span><span class="p">))</span>
<span class="nf">names</span><span class="p">(</span><span class="n">gl_kegg_list</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">as.character</span><span class="p">(</span><span class="n">gl_kegg</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">)</span>
<span class="n">gl_kegg_list</span><span class="o">&lt;-</span><span class="nf">sort</span><span class="p">(</span><span class="n">gl_kegg_list</span><span class="p">,</span> <span class="n">decreasing</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>

<span class="c1"># run GSEA of KEGG:</span>
<span class="n">KEGG_NK_Th</span><span class="o">&lt;-</span><span class="nf">gseKEGG</span><span class="p">(</span><span class="n">gl_kegg_list</span><span class="p">,</span> <span class="n">organism</span> <span class="o">=</span> <span class="s">&quot;hsa&quot;</span><span class="p">,</span> <span class="s">&quot;ncbi-geneid&quot;</span><span class="p">,</span>
                    <span class="n">minGSSize</span> <span class="o">=</span> <span class="m">30</span><span class="p">,</span>
                    <span class="n">eps</span><span class="o">=</span><span class="m">0</span><span class="p">,</span>
                    <span class="n">seed</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>
</code></pre></div>
<p>Explore the new object that was created. What is its structure? What does it contain? How many gene sets are up-regulated? Is their an immune-related gene set significant?
Because we had NK cells in our dataset, is the <a href="https://www.genome.jp/pathway/hsa04650">NK cell-mediated cytotoxicity</a> KEGG gene set significant? What is the total number of built-in
KEGG gene sets and how can you view the genes included in a particular gene set?</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code>    <span class="nf">str</span><span class="p">(</span><span class="n">KEGG_NK_Th</span><span class="p">)</span>
    <span class="nf">View</span><span class="p">(</span><span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">)</span>

    <span class="c1"># Up-regulated gene sets:</span>
    <span class="nf">summary</span><span class="p">(</span><span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">NES</span><span class="o">&gt;</span><span class="m">0</span><span class="p">)</span>

    <span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;immune&quot;</span><span class="p">,</span><span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="p">),</span> <span class="p">]</span>

    <span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;Natural killer&quot;</span><span class="p">,</span><span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">Description</span><span class="p">),</span> <span class="p">]</span>

    <span class="c1"># How many built-in KEGG pathways are there?</span>
    <span class="nf">length</span><span class="p">(</span><span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">geneSets</span><span class="p">)</span>
    <span class="c1"># Genes involved in Natural killer cell mediated cytotoxicity (hsa04650)</span>
    <span class="n">KEGG_NK_Th</span><span class="o">@</span><span class="n">geneSets</span><span class="o">$</span><span class="n">hsa04650</span> <span class="c1"># coded as Entrez Gene ID...</span>
</code></pre></div>
</details>
<p>Once you have performed a GSEA of the KEGG pathways, it is possible to overlay the fold change values of your genes on top of the
KEGG pathway maps available online. For this, create a named vector of fold change values, where the non-significant genes will be grey, while the up- and down-
regulated genes will be colored in the KEGG pathway map. The pathview() function comes from the pathview package (surprisingly&hellip;)</p>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">pathview</span><span class="p">)</span>

<span class="c1"># pathview map with non-significant genes in grey:</span>
<span class="c1"># set log fold change of non-significant genes to 0:</span>
<span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">logFC_0</span><span class="o">&lt;-</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">p.adj</span><span class="o">&gt;</span><span class="m">0.05</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">logFC</span><span class="p">)</span>

<span class="c1"># create named vector of fold change values:</span>
<span class="n">genePW</span><span class="o">&lt;-</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">logFC_0</span>
<span class="nf">names</span><span class="p">(</span><span class="n">genePW</span><span class="p">)</span><span class="o">&lt;-</span><span class="n">NK_vs_Th</span><span class="o">$</span><span class="n">symbol</span>

<span class="c1"># Create pathview map for Ribosome = hsa03010</span>
<span class="nf">pathview</span><span class="p">(</span><span class="n">gene.data</span>  <span class="o">=</span> <span class="n">genePW</span><span class="p">,</span>
         <span class="n">pathway.id</span> <span class="o">=</span> <span class="s">&quot;hsa03010&quot;</span><span class="p">,</span>
         <span class="n">species</span>    <span class="o">=</span> <span class="s">&quot;hsa&quot;</span><span class="p">,</span>
         <span class="n">gene.idtype</span> <span class="o">=</span> <span class="s">&quot;SYMBOL&quot;</span><span class="p">)</span>

<span class="c1"># Create pathview map of Natural killer cell mediated cytotoxicity = hsa04650</span>
<span class="nf">pathview</span><span class="p">(</span><span class="n">gene.data</span>  <span class="o">=</span> <span class="n">genePW</span><span class="p">,</span>
         <span class="n">pathway.id</span> <span class="o">=</span> <span class="s">&quot;hsa04650&quot;</span><span class="p">,</span>
         <span class="n">species</span>    <span class="o">=</span> <span class="s">&quot;hsa&quot;</span><span class="p">,</span>
         <span class="n">gene.idtype</span> <span class="o">=</span> <span class="s">&quot;SYMBOL&quot;</span><span class="p">)</span>
</code></pre></div>
<figure>
  <img src="../assets/images/hsa04650.pathview.png" width="700"/>
  </figure>

<p>Very often, researchers have their own gene sets in mind, either that they compiled from the literature, or that they defined from previous experiments. If this is the case for your experiment, you can generate an Excel file that contains one gene set per row, the 2 first columns contain the gene set ID and the gene set description, and the genes that belong to each gene set are listed in the next columns. To have an idea of
the structure of this file, open the .gmt file that we provided using Excel or a Text editor (NotePad, TextWrangler, BBEdit, etc).</p>
<p>In this exercise, we will perform a GSEA of the 50 hallmark gene sets that can be downloaded from <a href="http://www.gsea-msigdb.org/gsea/downloads.jsp">MSigDB</a>, which we can import using <code>clusterProfiler::read.gmt()</code>. 
Alternatively, we can import the Hallmark gene sets using <code>msigdbr</code>. What is the format of the object that is created? Use the GSEA() function, by providing the named and sorted
vector of t-statistics we used above with gseGO(). Why do we use this one and not the one created for gseKEGG()?</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Import hallmark, convert to term2gene and run GSEA:</span>
<span class="n">term2gene_h</span> <span class="o">&lt;-</span> <span class="nf">msigdbr</span><span class="p">(</span><span class="n">species</span> <span class="o">=</span> <span class="s">&quot;Homo sapiens&quot;</span><span class="p">,</span> <span class="n">category</span> <span class="o">=</span> <span class="s">&quot;H&quot;</span><span class="p">)</span>
<span class="c1"># Or alternatively:</span>
<span class="c1"># term2gene_h&lt;-read.gmt(&quot;h.all.v2023.2.Hs.symbols.gmt&quot;)</span>

<span class="nf">head</span><span class="p">(</span><span class="n">term2gene_h</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">term2gene_h</span><span class="o">$</span><span class="n">gs_name</span><span class="p">))</span> <span class="c1"># 50</span>

<span class="c1"># Run GSEA with the function that allows to use custom gene sets, </span>
<span class="c1"># provide the named vector of t statistics</span>
<span class="n">h_NK_vs_Th</span><span class="o">&lt;-</span><span class="nf">GSEA</span><span class="p">(</span><span class="n">gl</span><span class="p">,</span> <span class="n">TERM2GENE</span> <span class="o">=</span> <span class="n">term2gene_h</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gs_name&quot;</span><span class="p">,</span> <span class="s">&quot;gene_symbol&quot;</span><span class="p">)],</span>
                 <span class="n">eps</span><span class="o">=</span><span class="m">0</span><span class="p">,</span>
                 <span class="n">seed</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>

<span class="nf">View</span><span class="p">(</span><span class="n">h_NK_vs_Th</span><span class="o">@</span><span class="n">result</span><span class="p">)</span>

<span class="c1"># Number of significant gene sets:</span>
<span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">h_NK_vs_Th</span><span class="o">@</span><span class="n">result</span><span class="o">$</span><span class="n">p.adjust</span><span class="o">&lt;=</span><span class="m">0.05</span><span class="p">))</span> 
</code></pre></div>
<p>Finally, we can easily visualize the results:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># A dotplot with geneRatio or NES on the x-axis:</span>
<span class="nf">dotplot</span><span class="p">(</span><span class="n">h_NK_vs_Th</span><span class="p">)</span>
<span class="nf">dotplot</span><span class="p">(</span><span class="n">h_NK_vs_Th</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">&quot;NES&quot;</span><span class="p">,</span> <span class="n">orderBy</span><span class="o">=</span><span class="s">&quot;p.adjust&quot;</span><span class="p">)</span>

<span class="c1"># A barcode plot:</span>
<span class="nf">gseaplot2</span><span class="p">(</span><span class="n">h_NK_vs_Th</span><span class="p">,</span> <span class="n">geneSetID</span> <span class="o">=</span> <span class="s">&quot;HALLMARK_MTORC1_SIGNALING&quot;</span><span class="p">,</span>
           <span class="n">title</span><span class="o">=</span><span class="s">&quot;HALLMARK_MTORC1_SIGNALING&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="feedback">Feedback</h2>
<p>Thanks for attending this course! Don&rsquo;t forget to give honest feedback via the link sent by the course organizer.</p>
<h2 id="extra-exercise-for-ects-credits">Extra exercise for ECTS credits</h2>
<ul>
<li>Perform GSEA of the NK vs Th data using the Reactome gene sets either downloaded using the <code>msigdbr</code>  packege or downloaded from the 
MSigDB website <a href="http://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/2023.1.Hs/c2.cp.reactome.v2023.1.Hs.symbols.gmt">here</a>.</li>
<li>How many gene sets are significantly enriched?  </li>
<li>Generate a barplot of the NES of all significant gene sets, with the bars ordered according to NES value.  </li>
<li>Generate a barcode plot (i.e. gsea plot) for the gene set with the lowest NES.  </li>
</ul>
<p><strong>Steps</strong></p>
<ul>
<li>Import the reactome gene sets using <code>msigdbr()</code> with arguments <code>category="C2"</code> and <code>subcategory="CP:REACTOME"</code>, or import from the provided file <code>c2.cp.reactome.v2023.1.Hs.symbols.gmt</code> using <code>read.gmt()</code>.  </li>
<li>Use the <code>GSEA()</code> function providing a sorted named vector of t-statistics and the reactome gene sets, and use argument <code>minGSSize=30</code>.  </li>
<li>Count the number of significant adjusted p-values.  </li>
<li>Use <code>barplot()</code> and <code>gseaplot()</code> for the visualization of the results.  </li>
</ul>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        <a href="../materials/" class="md-footer__link md-footer__link--prev" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Materials
            </div>
          </div>
        </a>
      
      
        <a href="../bonus_code/" class="md-footer__link md-footer__link--next" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Bonus code
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../assets/javascripts/workers/search.d351de03.min.js", "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.34eae1b6.min.js"></script>
      
    
  </body>
</html>